---
title: "mapClinvarEdits"
author: "Martin Gordon"
date: "2024-08-06"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Map clinVar edits to CRISPR IDs

Want to map the edits to specifc CRISPR edits

```{r}
library(data.table)
library(magrittr)
library(ggplot2)
library(stringr)
```
read in the files
```{r}
clinvar <- fread('./data/GRCh38.snp_clinvar.txt')
clinvar <- clinvar[GeneSymbol %in% c('TSC1', 'TSC2', 'PIK3CA', 'PDK1', 'AKT1', 'MTOR')]
# subset to the cols we want
clinvar <- clinvar[,.(AlleleID=`#AlleleID`, Name, GeneSymbol, HGNC_ID, ClinicalSignificance, ReviewStatus, dbSNP=`RS# (dbSNP)`, ChromosomeAccession, Chromosome, Start, Stop, ReferenceAlleleVCF, AlternateAlleleVCF)]

crispr.dt  <- lapply(dir('./data', full.names = T, pattern='*guides*'), function(x){
  dt <- fread(x)
  dt[, editor := str_extract(x,'Abe8e|BEmax')]
}) %>%  rbindlist()
```

inspect the data
```{r}
# Transcript and reference alleles dont match for genes on the -ve strand (relative to reference)
crispr.dt[`Transcript reference allele` != `Genome reference allele`,.N, by=.(`Gene strand`, `sgRNA Strand`)]

# edit window is 4-8 for both
#or Abe8e this is a A-G edit, 
# for BEMAx, this is CT
crispr.dt %>%  str()

crispr.dt[, head(.SD, 10), by = c("editor","Gene strand", "sgRNA Strand")]

# what does this make the edits look like?
# first step:is sgRNA sense or antisense? if sense then the trqnscript edit is simple (A-G or C-T dependent on editor), if the sgRNA is antisense, then T-C or G-A (outcome of the edit on the produced transcript)
# second: is the gene on ++ or - strand. If + strand; then the reference ref and alt are the same as the transcript edit (produced from the same strand), if gene on -ve strand, then reverse complement of the edit

# can see only 8 possible combinations here
crispr.dt[,.(editor,`Gene strand`,`sgRNA Strand`,`Transcript reference allele`, `Transcript alternate allele`,`Genome reference allele`,`Genome alternate allele`)] %>% 
  unique()

crispr.dt[`sgRNA sequence` ==  'AGCGAGAGAATGGCCCAACA']


crispr.dt$`Ensembl transcript ID` %>%  unique()

# check  
```




```{r}

clinvar <- clinvar[,.(AlleleID=`#AlleleID`, Name, GeneSymbol, HGNC_ID, ClinicalSignificance, ReviewStatus, dbSNP=`RS# (dbSNP)`, ChromosomeAccession, Chromosome, Start, Stop, ReferenceAlleleVCF, AlternateAlleleVCF)]
```


