---
title: "050225_kroganWeeklyPlots"
author: "Martin Gordon"
date: "2025-05-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## overview
For now we just want to prepare plots for Ronald for next Thursdays weekly meeting

```{r packages}
library(data.table)
library(ggplot2)
library(magrittr)
library(ggrepel)
library(stringr)
library(ComplexHeatmap)
library(circlize)
library(viridis)
library(RColorBrewer)
library(hrbrthemes) # visually nice set of themes
library(patchwork)
library(showtext)
library(seqinr)
library(readxl)
library(ggridges)
library(RColorBrewer)
library(ggforce)
library(stringr)
library(ggnewscale) # multiple aes sclaes
library(gggenes)

source("../../utils/bp_utils/MSstats_Helper_Functions.R")
source("../../utils/bp_utils/ManageScriptData.R")
source ("../../utils/bp_utils/enrichmentTestFunctions.R")
source ("../../utils/mg_utils/r_utils/IDmapping.R")
source ("../../utils/mg_utils/r_utils/plottingHelperFunctions.R")
source("../../utils/mg_utils/r_utils/CRISPR_Functions.R")


# function to cluster w NA values
clusterWNA <-  function(mat, na.value=0,...){
  
  mat[is.na(mat)] <- na.value
  return(hclust(dist(mat)))
}

customTheme <- theme_bw() +
  theme(panel.border = element_rect(color = "lightgrey", 
                                    fill = NA, 
                                    size = 1),
        axis.text.x = element_text(angle=90)
        )

#set one
col.pal <- getQualitativePalette(n=13)
#col.pal <- randomcoloR::distinctColorPalette(k=13)

# for heatmaps etc use this palette
redBlue <- RedBlueColPal()

# clinvar palette
clinvar.pal <- c(brewer.pal(8, "Blues"), 'grey')

domain.pal <- c('PH'= "#CC71C4" ,'Protein kinase'="#DBDB58","AGC-kinase C-terminal"="#8246DB","FAT"="#E1575B","PI3K/PI4K catalytic"="#D2E2A8",
                "FATC"="#D2AEDB","PI3K-ABD"="#80D5CB","PI3K-RBD"="#D1919A", "C2 PI3K-type"="#7C88DB", "PIK helical"="#77DC97", "PI3K/PI4K catalytic"="#D5A261",
                "Phosphatase tensin-type"= "#77B2D4", "C2 tensin-type"="#ffd8b1", "grey90"="na")
```

```{r tables}
counts.dt <- fread('~/Documents/projects/101224_RBabu_CRISPRe_PIK3CA/012825_requestedMeetingPlots_data/2025_01_29_sampleCounts.rawAndNormalized.dt')
counts.dt$group %>% unique()
grp.levels <- c('LibraryRepresentation_0', 'Alpelisib_22', 'Alpelisib_7', 'DMSO_22', 'DMSO_7', 'Paxalisib_22', 'Paxalisib_7')
counts.dt[, group := factor(paste(treatment, timepoint, sep='_'), levels=grp.levels)]

# also nned the PW comparisons and anno; this should be the set of everything we have
de.dt <- fread('~/Documents/projects/101224_RBabu_CRISPRe_PIK3CA/111824_plotsForKroganWeekly_data/2025_01_22_deseq.pwcomparisons.allAnno.csv')

gene.oi <- c('MTOR','PIK3CA', 'PTEN', 'AKT1')

de.dt[, c('numerator', 'denominator') := tstrsplit(contrast, '_vs_', keep=c(1,2))]
de.dt[, numerator := factor(numerator, levels=c('DMSO_7','DMSO_22', 'Paxalisib_7','Paxalisib_22', 'Alpelisib_7','Alpelisib_22'))]
de.dt[, numeratorTreatment := gsub('_[0-9]{1,2}', '', numerator)]
de.dt[, Gene := gene] # just to avoid reediting


selected.targets <- setDT(read_xlsx('./data/ronald_targets.xlsx', col_names =F))
setnames(selected.targets, new=c('guideName', 'sequence','status'))

de.dt <- merge(de.dt, selected.targets[,.(guideName,sequence, status)], by.x=c('guide'), by.y=c('sequence'), all.x=T)

# get the list of IDs to label
guidesOI <- de.dt[status != 'NA', unique(id)]
```

```{r col-palette}

brewer.pal(n=9, name = 'Blues')[4:8]
brewer.pal(n=9, name = 'PuRd')[4:8]
brewer.pal(n=9, name = 'Reds')[4:8]
brewer.pal(n=9, name = 'YlOrRd')[4:8]
brewer.pal(n=9, name = 'YlGn')[4:8]
brewer.pal(n=9, name = 'Greys')[4:8]
# blues

group.pal <- c("#BDBDBD", "#DF65B0", "#C994C7", "#6BAED6", "#9ECAE1","#FB6A4A", "#FC9272" )
#names(group.pal) <- unique(counts.dt$group)
names(group.pal) <- c('LibraryRepresentation_0', 'Alpelisib_22', 'Alpelisib_7', 'DMSO_22 DMSO_7', 'Paxalisib_22', 'Paxalisib_7')

# "#FC9272" "#FB6A4A" "#EF3B2C" "#CB181D" "#A50F15"
# "#9ECAE1" "#6BAED6" "#4292C6" "#2171B5" "#08519C"
# "#C994C7" "#DF65B0" "#E7298A" "#CE1256" "#980043"
# "#ADDD8E" "#78C679" "#41AB5D" "#238443" "#006837"
# "#BDBDBD" "#969696" "#737373" "#525252" "#252525"
# "#FEB24C" "#FD8D3C" "#FC4E2A" "#E31A1C" "#BD0026"


# timepoint and editor col pal
counts.dt$timepoint %>% unique()

group.pal <- c("#BDBDBD", "#DF65B0", "#C994C7", "#6BAED6", "#9ECAE1","#FB6A4A", "#FC9272" )
names(group.pal) <- unique(counts.dt$group)

timepoint.pal <- c("#ADDD8E" ,"#78C679", "#41AB5D")
names(timepoint.pal) <-c('0','7', '22')

editor.pal <- brewer.pal(n=9, name = 'Set3')[4:6]
names(editor.pal) <- c('abe8e', 'bemax', 'plasmid')

gene.col <- brewer.pal(n=6, name = 'Set2')
# set6 if need non targeting
gene.col <- c("#66C2A5" ,"#FC8D62","#FFD92F", "#8DA0CB") # "#A6D854" "#FFD92F" "#E78AC3""#8DA0CB"



# uniprot Ids PTEN, MTOR,  AKT1, PIK3CA,
uniprot.ids <- c('P60484', 'P42345', 'P31749', 'P42336')
```
Read in the alphamissense results, subset to the genes we want, look at the mutation breakdown and then
```{r}
ams.dt <- fread('/Users/martingordon/Library/CloudStorage/Box-Box/AlphaMissense/AlphaMissense_aa_substitutions.tsv.gz')
ams.dt <- ams.dt[uniprot_id %in% uniprot.ids]

# add gene names 
ams.dt[, gene := ifelse(uniprot_id == 'P31749', 'AKT1',
                        ifelse(uniprot_id == 'P60484', 'PTEN',
                               ifelse(uniprot_id == 'P42345', 'MTOR',
                                      'PIK3CA')))]

ams.dt[, pos := gsub('[A-Z]', '', protein_variant)]
ams.dt[, c('ref', 'alt') := tstrsplit(protein_variant, '[0-9]+', keep=c(1,2))]
ams.dt[, pos := as.integer(pos)]
```
Make abe8e long format and facet

need to convert ref_peptide, alt_peptide and peptide_position to be the same naming convention for splitting; just rename as now in the code so all are na or all are empty
```{r}
# simple fix I think.. looks good on inspection
de.dt[, peptide_position_simplified := gsub('NA', '', peptide_position)]

expandcols <- c('ref_peptide', 'alt_peptide','peptide_position_simplified')
cols.oi <- c('guide', 'editor', 'id', 'gene', 'pep_mutant_id','log2FoldChange', 'pvalue', 'padj', 'contrast', 'sig', 'numeratorTreatment', 'numeratorTimepoint', 'status')
genes.oi <- c("MTOR", "AKT1","PIK3CA","PTEN")

expand.dt <- de.dt[, lapply(.SD, function(x) unlist(tstrsplit(x, ';'))), .SDcols=expandcols, by=c(cols.oi)]

#fwrite(expand.dt, ScriptAndDatedFileName('expandedGuides.deseq.pwcomparisons.csv.gz'))
expand.dt <- fread('~/Documents/projects/101224_RBabu_CRISPRe_PIK3CA/050225_kroganWeeklyPlots_data/2025_05_05_expandedGuides.deseq.pwcomparisons.csv.gz')
expand.dt[, peptide_position_simplified := as.integer(peptide_position_simplified)]
```

Combine with AF missense and plot 
Overlap with ref residue is great! the missing are the edits that introduce silent mutations
```{r}
# find how to combine
# ~ 170k
expand.dt <- expand.dt[!(alt_peptide %in% c('*', ''))] 
expand.dt <- expand.dt[!(ref_peptide %in% c('*', ''))] 
expand.dt[ref_peptide != alt_peptide] # and noself edits/hits...

# lost many guides... need to figure out what is missing.. shouldnt we have all sites?
# ensure the translated codons match what is expected

expand.dt <- merge(x=expand.dt, y=ams.dt[, .(gene, ref, pos, alt, am_class, am_pathogenicity, am_ref=ref, am_alt=alt, am_pos=pos)], by.x=c('gene', 'ref_peptide', 'alt_peptide','peptide_position_simplified'), by.y=c('gene', 'ref', 'alt', 'pos'))

#fwrite(expand.dt, ScriptAndDatedFileName('expanded.deseq.pwcomparisons.alphamissenseAnno.csv.gz'))
expand.dt <- fread('~/Documents/projects/101224_RBabu_CRISPRe_PIK3CA/050225_kroganWeeklyPlots_data/2025_05_05_expanded.deseq.pwcomparisons.alphamissenseAnno.csv.gz')
```

Look at the distribution of pathogenicity scores for each of the genes; 
bimodal mixture
Go with the class labels for now, but I think we might want to set our own values
```{r}
ggplot(ams.dt, aes(x=am_pathogenicity)) +
  geom_histogram(bins=100) +
  facet_grid(gene~.)

# seems to be some score threshold that differentiates these.. not sure if I want to use c
ggplot(ams.dt, aes(x=am_class, y=am_pathogenicity, color=am_class)) +
  geom_sina()
```

```{r}
expand.dt[, am_class := factor(am_class, levels=c('benign', 'ambiguous', 'pathogenic'))]

ggplot(expand.dt, aes(x=am_class, y=log2FoldChange, fill=sig)) +
  geom_boxplot() +
  facet_wrap(gene~editor)
expand.dt[editor == 'bemax' & gene == 'PTEN' & sig == 'up']
```
This is very interesting; we can clearly see that there seems to be some enrichment for the disruptive mutations on the right.. Want to define different groups and see how the enriched we are for significant genes in each. set

```{r}
ggplot(expand.dt[contrast =='Alpelisib_22_vs_LibraryRep_0' & editor == 'abe8e'], aes(x=am_pathogenicity, y=-log10(pvalue), color=sig)) +
  geom_point() +
  facet_wrap(~gene) +
  geom_vline(xintercept=0.75)
```
Look at pathogenicity scores of Ronalds selected guides; only 23 out of the 27 are recovered as the others seem to be splice acceptors and/or a silent mutation
```{r}
labelSimplified <- expand.dt[status != '', .(id, am_class, pep_mutant_id,  am_pathogenicity)] %>% 
  .[, .SD[which.max(am_pathogenicity)], by=id]

# one missing is 1050.... others are slpice sites. weird snps to choose?
de.dt[id %in% guidesOI[!guidesOI %in% labelSimplified$id] & peptide_position == '1050',]
ams.dt[gene =='PIK3CA' & pos == 1050]
```

First things first, replot the heatmaps I previously shared with Ronald of his guides, improve the label names, remove the redundancy and add the clinvarAnnotations
Merge the counts dt with the annotation information and plot
```{r}
counts.dt[de.dt[, .(sgRNA=id, editor, pep_mutant_id)], mutantID := i.pep_mutant_id, on=c('sgRNA', 'editor')]

counts.dt[, hmID :=ifelse(is.na(mutantID) | mutantID == '::', paste0(gene,'_', sgRNA), paste0(gene, '__', mutantID))]

annodt <- counts.dt[sgRNA %in% guidesOI, .(gene, editor, sgRNA, hmID)] %>% 
  .[editor %in% c('abe8e', 'bemax')] %>% 
  unique()

# merge the pathogenicity info to the table
annodt <- merge(annodt, labelSimplified[,.(id,am_class, am_pathogenicity)], by.x=c('sgRNA'), by.y='id', all.x=T)
# how to update multiple columns at once
annodt[is.na(am_class), names(.SD) := lapply(.SD, function(x) ifelse(!is.na(x), x, NA)), .SDcols = c('am_class', 'am_pathogenicity')]

```


plot the heatmaps with the AF missense pathogenicity
```{r}

abe8e.mat <- dcast(counts.dt[editor == 'abe8e' & sgRNA %in% guidesOI,], hmID~paste0(treatment, '_', timepoint,'_', replicate), value.var='sgRNAnormCounts') %>% 
  as.matrix(rownames=1)
bemax.mat <- dcast(counts.dt[editor == 'bemax' & sgRNA %in% guidesOI,], hmID~paste0(treatment, '_', timepoint,'_', replicate), value.var='sgRNAnormCounts') %>% 
  as.matrix(rownames=1)

# sweep the library Rep valuea
bemax.mat <- log2(bemax.mat +1)
abe8e.mat <- log2(abe8e.mat +1)

bemax.sub <- sweep(bemax.mat, 1, apply(bemax.mat[, grepl('LibraryRepresentation', colnames(bemax.mat))], 1, mean))
abe8e.sub <- sweep(abe8e.mat, 1, apply(abe8e.mat[, grepl('LibraryRepresentation', colnames(abe8e.mat))], 1, mean))

bemax.sub <- bemax.sub[, !grepl('Library', colnames(bemax.sub))]
abe8e.sub <- abe8e.sub[, !grepl('Library', colnames(abe8e.sub))]

corder <- c("DMSO_7_1","DMSO_7_2", "DMSO_22_1","DMSO_22_2","Paxalisib_7_1","Paxalisib_7_2",
            "Paxalisib_22_1","Paxalisib_22_2","Alpelisib_7_1","Alpelisib_7_2","Alpelisib_22_1","Alpelisib_22_2")

# sort row and col order
bemax.sub <- bemax.sub[,corder]
abe8e.sub <- abe8e.sub[,corder]
# reorder mat to match annodt
bemax.sub <- bemax.sub[match(annodt[editor=='bemax', hmID], rownames(bemax.sub)), ]
abe8e.sub <- abe8e.sub[match(annodt[editor=='abe8e', hmID],rownames(abe8e.sub)), ]
  

names(gene.col) <- unique(annodt$gene)
annodt
# rowHa <- rowAnnotation('gene'=annodt[editor == 'bemax', gene], 'AlphaMissense\npathogenicity'=annodt[editor == 'bemax',am_pathogenicity],
#                        col=list('gene'=gene.col),
#                        border=T)
rowHa <- rowAnnotation('AlphaMissense\npathogenicity'=annodt[editor == 'bemax',am_pathogenicity],
                       border=T)

hm <- Heatmap(bemax.sub, 
              cluster_columns = F,
              cluster_rows=F, 
              row_split = annodt[editor == 'bemax',gene],
              show_column_names = F,
              cluster_column_slices = F,
              col=colorRamp2(breaks=c(-2,0,2), colors = c(redBlue[5],  redBlue[3], redBlue[1])),
              #column_split = factor(gsub('_[12]','', corder), levels = unique(sub('_[12]','', corder))),
              column_split = factor(gsub("_[12]{1}$", '', colnames(bemax.sub)), levels=c('DMSO_7', 'DMSO_22', 'Paxalisib_7', 'Paxalisib_22', 'Alpelisib_7', 'Alpelisib_22')),
              name='log2 counts/\nmean(Librep)',
              border=T,
              row_names_gp = gpar(fontsize=8),
              right_annotation = rowHa,
              row_title_gp=gpar(fontsize=10, fontface = "bold"),
              column_title_gp=gpar(fontsize=12, fontface = "bold"),
              show_row_names = T)
hm
BackupAsPDF(hm, 'bemaxEditor.selectedGuides.heatmap', dimensions=c(10,7))

rowHa <- rowAnnotation( 'AlphaMissense\npathogenicity'=annodt[editor == 'abe8e',am_pathogenicity],
                       border=T)

hm <- Heatmap(abe8e.sub, 
              cluster_columns = F,
              cluster_rows=F, 
              row_split = annodt[editor == 'abe8e',gene],
              show_column_names = F,
              cluster_column_slices = F,
              col=colorRamp2(breaks=c(-4,0,4), colors = c(redBlue[5],  redBlue[3], redBlue[1])),
              #column_split = factor(gsub('_[12]','', corder), levels = unique(sub('_[12]','', corder))),
              column_split = factor(gsub("_[12]{1}$", '', colnames(abe8e.sub)), levels=c('DMSO_7', 'DMSO_22', 'Paxalisib_7', 'Paxalisib_22', 'Alpelisib_7', 'Alpelisib_22')),
              name='log2 counts/\nmean(Librep)',
              border=T,
              right_annotation = rowHa,
              row_title_gp=gpar(fontsize=10, fontface = "bold"),
              column_title_gp=gpar(fontsize=12, fontface = "bold"),
              show_row_names = T)
hm
BackupAsPDF(hm, 'abe8eEditor.selectedGuides.heatmap', dimensions=c(10,7))
```
Quick thing to do; for each of our annotations; just plotted one see loop over all below
```{r}
expand.dt[, sigLabel := ifelse(abs(log2FoldChange) >=1 & pvalue < 0.005, 'yes', 'no')]
cols.oi <- c('sigLabel', 'set')
counts.dt[, c(cols.oi) := lapply(.SD, as.factor), .SDcols = cols.oi]
# prepare this for fishers exact test
expand.dt[, AMlab := ifelse(am_class == 'pathogenic', 'pathogenic', 'benign/uncertain')]

pdf('~/Documents/projects/101224_RBabu_CRISPRe_PIK3CA/050225_kroganWeeklyPlots_data/pdfs//Paxalisib_22_vs_LibraryRep_0.mosaicplot2025_05_02__22_14__08.00.pdf')
g <- mosaicplot(xtabs(~sigLabel+am_class, expand.dt[contrast=='Paxalisib_22_vs_LibraryRep_0']),
           color = c( 'cornflowerblue','grey70', 'firebrick2', 'cornflowerblue','grey70', 'firebrick2'),
           main = 'Paxalisib_22_vs_LibraryRep_0',
           margin=c(10,20),
           off = c(2,5))
dev.off()
```
**Fishers Exact test**
Run a fishers exact test on each of the contrasts; see if we are significantly enriched over bg

```{r}
# just convert to a simple binary choice
expand.dt[, AMlab := ifelse(am_class == 'pathogenic', 'pathogenic', 'benign/uncertain')]

# for rowwise selection of groups, use this for enrihcment testing
new.fisher.results <- expand.dt[ , 
  if (.N > 0L) {
    cont <- table(sigLabel, AMlab)
    if (all(is.finite(cont)) && all(cont >= 0) && nrow(cont) > 1 && ncol(cont) > 1) {
      fisher.p = fisher.test(cont, alternative='greater')$p.value
    } else {
      fisher.p = 1 # another option??
    }
  }, by = .(contrast, gene)]

new.fisher.results[, padj := p.adjust(V1, method='BH')]
setnames(new.fisher.results, old='V1', new='pvalue')
fwrite(new.fisher.results, ScriptAndDatedFileName('fishersTestresults.alphamissense.enrichment.csv'))


# this method would be if you want to test multiple columns! Perhaps could use with a matrix input and patterns!!
lm_coef = sapply(contrasts.oi, function(rhs) {
  
  expand.dt[ , fisher.test(table(sigLabel, AMlab))$p.value, .SDcols = c('gene', rhs)]
})

g <- ggplot(new.fisher.results[grepl('_22_vs', contrast)], aes(x=gene, y=-log10(V1), fill=gene)) +
  geom_hline(yintercept = -log10(0.05), linetype=2) +
  geom_bar(stat='identity') +
  facet_wrap(~contrast) +
  labs('results of fishers exact test') +
  scale_fill_manual(values=unname(gene.col)) +
  theme_bw() +
  customTheme
g
BackupAsPDF(g, 'fishers.test.res.barplot')

```
plot mosaic plots for all the genes and contrasts seperately and show that. Do the same for clinvar


```{r, heatmap, fig.height=7, fig.width=8}
expand.dt$contrast %>% 
  unique()

contrasts.oi <- c('Paxalisib_22_vs_LibraryRep_0', 'Alpelisib_22_vs_LibraryRep_0', 'DMSO_22_vs_LibraryRep_0', 'Paxalisib_7_vs_LibraryRep_0', 'Alpelisib_7_vs_LibraryRep_0', 'DMSO_7_vs_LibraryRep_0')

g <- ggplot(expand.dt[sig != 'not' & contrast %in% contrasts.oi,], aes(x=am_class, y=abs(log2FoldChange), color=am_class)) +
  geom_boxplot(outliers=F, show.legend = F) +
  geom_point(aes(fill = am_class), size = 1.25, shape = 21, alpha=0.7, position = position_jitterdodge(jitter.width=0.2, dodge.width = .75, seed=1), color='black') +
  geom_text_repel(data=expand.dt[sig != 'not' & contrast %in% contrasts.oi & status != '',], aes(label=pep_mutant_id),
                  position = position_jitterdodge(jitter.width=0.2, dodge.width = .75, seed=1), size=1.5) +
  ggtitle('Fold change distributions', subtitle='Grouped by AlphaMisense pathogenicity score') +
  coord_flip() +
  facet_grid(numeratorTreatment~numeratorTimepoint) +
  scale_color_manual(values=c(pathogenic=col.pal[2], ambiguous='grey70', benign='cornflowerblue')) +
  scale_fill_manual(values=c(pathogenic=col.pal[2], ambiguous='grey70', benign='cornflowerblue')) +
  theme_bw() +
  customTheme +
  guides(fill="none", 
         color="none")
g
BackupAsPDF(g, 'alphaMissense.lfc.distributions.boxplots')


g <- ggplot(expand.dt[sig != 'not' & contrast %in% contrasts.oi,], aes(x=am_class, y=abs(log2FoldChange), color=am_class)) +
  geom_sina(alpha=0.7) +
  #geom_point(aes(fill = am_class), size = 1.25, shape = 21, alpha=0.7, position = position_jitterdodge(jitter.width=0.2, dodge.width = .75, seed=1), color='black') +
  geom_text_repel(data=expand.dt[sig != 'not' & contrast %in% contrasts.oi & status != '',], aes(label=pep_mutant_id),
                  position = position_jitterdodge(jitter.width=0.2, dodge.width = .75, seed=1), size=1.5) +
  ggtitle('Fold change distributions', subtitle='Grouped by AlphaMisense pathogenicity score') +
  coord_flip() +
  facet_grid(numeratorTreatment~numeratorTimepoint) +
  scale_color_manual(values=c(pathogenic=col.pal[2], ambiguous='grey70', benign='cornflowerblue')) +
  scale_fill_manual(values=c(pathogenic=col.pal[2], ambiguous='grey70', benign='cornflowerblue')) +
  theme_bw() +
  customTheme +
  guides(fill="none", 
         color="none")
g
BackupAsPDF(g, 'alphaMissense.lfc.distributions.sina')

# facet by time
g <- ggplot(expand.dt[sig != 'not' & contrast %in% contrasts.oi,], aes(x=am_class, y=abs(log2FoldChange), color=am_class)) +
  geom_sina(alpha=0.7) +
  #geom_point(aes(fill = am_class), size = 1.25, shape = 21, alpha=0.7, position = position_jitterdodge(jitter.width=0.2, dodge.width = .75, seed=1), color='black') +
  geom_text_repel(data=expand.dt[sig != 'not' & contrast %in% contrasts.oi & status != '',], aes(label=pep_mutant_id),
                  position = position_jitterdodge(jitter.width=0.2, dodge.width = .75, seed=1), size=1.5) +
  ggtitle('Fold change distributions', subtitle='Grouped by AlphaMisense pathogenicity score') +
  coord_flip() +
  facet_grid(numeratorTreatment~numeratorTimepoint) +
  scale_color_manual(values=c(pathogenic=col.pal[2], ambiguous='grey70', benign='cornflowerblue')) +
  scale_fill_manual(values=c(pathogenic=col.pal[2], ambiguous='grey70', benign='cornflowerblue')) +
  theme_bw() +
  customTheme +
  guides(fill="none", 
         color="none")
g
BackupAsPDF(g, 'alphaMissense.lfc.distributions.sina')


# facet by time
g <- ggplot(expand.dt[sig != 'not' & contrast %in% contrasts.oi,], aes(x=am_class, y=abs(log2FoldChange), color=am_class)) +
  geom_sina(alpha=0.7) +
  #geom_point(aes(fill = am_class), size = 1.25, shape = 21, alpha=0.7, position = position_jitterdodge(jitter.width=0.2, dodge.width = .75, seed=1), color='black') +
  #geom_text_repel(data=expand.dt[sig != 'not' & contrast %in% contrasts.oi & status != '',], aes(label=pep_mutant_id),
  #                position = position_jitterdodge(jitter.width=0.2, dodge.width = .75, seed=1), size=1.5) +
  ggtitle('Fold change distributions', subtitle='Grouped by AlphaMisense pathogenicity score') +
  coord_flip() +
  facet_grid(numeratorTreatment~numeratorTimepoint) +
  scale_color_manual(values=c(pathogenic=col.pal[2], ambiguous='grey70', benign='cornflowerblue')) +
  scale_fill_manual(values=c(pathogenic=col.pal[2], ambiguous='grey70', benign='cornflowerblue')) +
  theme_bw() +
  customTheme +
  guides(fill="none", 
         color="none")
g
BackupAsPDF(g, 'alphaMissense.lfc.distributions.tp7and22.sina')
```
show the breakdown of DE genes in ggplot mosaic plot for both clinvar and Amissense
For now we are using our significance threshold to differentiate
```{r}
summary.dt <- expand.dt[, .(count=.N), by=.(contrast, am_class, sigLabel)]
summary.dt[, total := sum(count,na.rm = T), by=.(contrast, sigLabel)]

summary.dt$am_class %>% unique()
cols.oi <- c('sigLabel', 'am_class')
summary.dt[, names(.SD) := lapply(.SD, as.factor), .SDcols=cols.oi]
summary.dt[, sigLabel := factor(sigLabel, levels=c('no','yes'))]
summary.dt[, am_class := factor(am_class, levels=c('pathogenic', 'ambiguous', 'benign'))]
```

test plot; loop over all the contrasts and plot this; then break this down for a subset of contrasts and their genes
```{r}
# just use gplot
g <- ggplot(summary.dt[contrast =='Alpelisib_22_vs_DMSO_22',], aes(x=sigLabel, y=count, width=c(0.9,0.9,0.9,0.5,0.5,0.5),fill=am_class)) +
  geom_col(color='black', width=0.05, position='fill') +
  #geom_label(data=summary.dt[contrast =='Alpelisib_22_vs_DMSO_22',], aes(y=ylab), color='black', fill='white', size=4) +
  geom_label(data=summary.dt[contrast == 'Alpelisib_22_vs_DMSO_22',.(total=sum(count)), by=sigLabel], aes(y=1+.05, x=sigLabel, label=paste0('N=',as.character(total))), inherit.aes = F) +
  facet_grid(~gene, scales = "free_x", space = "free_x")
  scale_fill_manual(values=c(pathogenic=col.pal[2], ambiguous='grey70', benign='cornflowerblue')) + 
  labs(title='contrast') +
  theme_void() +
  guides(fill=guide_legend(title='AlphaMissense classification'))
g
```
```{r, mosaic-plot, fig.height=4, fig.width=6}
contrasts.oi <- unique(summary.dt$contrast)

summary.dt <- summary.dt[order(sigLabel)]

lapply(contrasts.oi, function(x){
  
  col.width <- c(0.9,0.9,0.9,0.5,0.5,0.5)
  
  g <- ggplot(summary.dt[contrast == x,], aes(x=sigLabel, y=count, width=col.width, fill=am_class)) +
  geom_col(color='black', width=0.05, position='fill') +
  #geom_label(data=summary.dt[contrast =='Alpelisib_22_vs_DMSO_22',], aes(y=ylab), color='black', fill='white', size=4) +
  geom_label(data=summary.dt[contrast == x,.(total=sum(count)), by=sigLabel], aes(y=1+.05, x=sigLabel, label=paste0('N=',as.character(total))), inherit.aes = F) +
  #facet_grid(~gene, scales = "free_x", space = "free_x")
  scale_fill_manual(values=c(pathogenic=col.pal[2], ambiguous='grey70', benign='cornflowerblue')) + 
  labs(title=x, y='Proportion', x='|LFC| > 1 & pvalue < 0.005') +
  theme_classic() +
  guides(fill=guide_legend(title='AlphaMissense classification'))
  
  g
  BackupAsPDF(g, paste0('alphamissense_mosaic/',x,'.mosaic'))
})
```
read in the guides with the clinvar annotations and we can join to the datatable for this
One issue with this is we are not matching to non-coding; for now lets just look at enrichments and proporitons
```{r}
#all the expanded guides
expand.dt <- fread('~/Documents/projects/101224_RBabu_CRISPRe_PIK3CA/050225_kroganWeeklyPlots_data/2025_05_05_expandedGuides.deseq.pwcomparisons.csv.gz')

# read in the clinvar annotated version
# read in a clinvar guide annotation, and use this to merge to the dt with most severe variant attached
clinvar.anno.long <- fread('/Users/martingordon/Documents/projects/101224_RBabu_CRISPRe_PIK3CA/110724_annotateCRISPRguides_Functions_data/2024_11_14_guides.clinvar.anno.long.txt')

# want cols for filtering to be explicitly missing
clinvar.anno.long[ref_prot == '', ref_prot := NA_character_]
clinvar.anno.long[alt_prot == '', alt_prot := NA_character_]

#' simple function to convert between standard one leeter and 3 letter peptide symbols
#' based on https://stackoverflow.com/questions/34296400/1-letter-amino-acid-variant-to-3-letter
convertAASymbol <- function(x, aaType='oneLetter'){
  
  # if empty return empty
  if (is.na(x)){
    return(NA_character_)
  }
  stopifnot(aaType %in% c('oneLetter', 'threeLetter'))
  if ((nchar(x) != 1 && aaType == 'oneLetter') ||
      (nchar(x) != 3 && aaType == 'threeLetter')) {
    stop("Input length must be 1 for 'oneLetter' or 3 for 'threeLetter'")
  }
  aaVec <- c(
                    "A"="Ala", "R"="Arg", "N"="Asn", "D"="Asp",
                    "C"="Cys", "E"="Glu", "Q"="Gln", "G"="Gly",
                    "H"="His", "I"="Ile", "L"="Leu", "K"="Lys",
                    "M"="Met", "F"="Phe", "P"="Pro", "S"="Ser",
                    "T"="Thr", "W"="Trp", "Y"="Tyr", "V"="Val",
                    "*"="Ter"
                    ) 
  
  if (aaType == 'threeLetter'){
    conversionVec <- names(aaVec)
    names(conversionVec) <- aaVec
  } else {
    conversionVec <- aaVec
  }
  
  if (x  %in% names(conversionVec)){
    return(stringr::str_replace_all(x, conversionVec))
  }
    message(paste0(x, ' not found in aa table. Returning NULL..'))
    return(NA_character_)
}

cols.oi <- c('guide', 'gene', 'editor','ref_peptide', 'alt_peptide','hgvs', 'clinicalSignificance', 'prot_site')

#
clinvar.anno.long[, ref_peptide := convertAASymbol(ref_prot, aaType='threeLetter'), by=.I]
clinvar.anno.long[, alt_peptide := convertAASymbol(alt_prot, aaType='threeLetter'), by=.I]

clinvar.anno.long

clinvar.expand <- merge(expand.dt, clinvar.anno.long[,..cols.oi], by.x=c('guide', 'gene', 'editor','ref_peptide', 'alt_peptide',  'peptide_position_simplified'),  by.y=c('guide', 'gene', 'editor', 'ref_peptide', 'alt_peptide', 'prot_site'),all.x=T)

#fwrite(clinvar.expand, ScriptAndDatedFileName('expanded.deseq.pwcomparisons.clinvarAnno.csv.gz'))
clinvar.expand <- fread('~/Documents/projects/101224_RBabu_CRISPRe_PIK3CA/050225_kroganWeeklyPlots_data/2025_05_05_expanded.deseq.pwcomparisons.clinvarAnno.csv.gz')
```

# set labels on clinvar expanded;
For fisher test, we need two clean labels; either pathogenic or not
```{r}
clinvar.expand[, sigLabel := ifelse(abs(log2FoldChange) >= 1 & pvalue < 0.005, 'yes', 'no')]

# subset to simplier label set
clinvar.oi <- c('Likely benign', 'Benign', 'Uncertain significance', 'Pathogenic', 'Likely pathogenic', 'Benign/Likely benign', 'Pathogenic/Likely pathogenic')
clinvar.expand <- clinvar.expand[clinicalSignificance %in% clinvar.oi,]

clinvar.label <- grep('patho', clinvar.expand$clinicalSignificance, value = T, ignore.case = T) %>% 
  unique()

clinvar.expand[, clinvarLabel := 'Benign/Likely benign']
clinvar.expand[clinicalSignificance %in% clinvar.label,  clinvarLabel := 'Pathogenic/Likely pathogenic']
clinvar.expand[clinicalSignificance == 'Uncertain significance', clinvarLabel := 'Uncertain significance']

# sanity check 
clinvar.expand[,.N, by=.(clinvarLabel,clinicalSignificance)]
```
First lets plot the mosaic plots
```{r}
summary.dt <- clinvar.expand[, .(count=.N), by=.(contrast, clinvarLabel, sigLabel)]
summary.dt[, total := sum(count,na.rm = T), by=.(contrast, sigLabel)]
summary.dt[contrast == 'Alpelisib_22_vs_LibraryRep_0']

cols.oi <- c('sigLabel', 'clinvarLabel')
summary.dt[, names(.SD) := lapply(.SD, as.factor), .SDcols=cols.oi]
summary.dt[, sigLabel := factor(sigLabel, levels=c('no','yes'))]
summary.dt[, clinvarLabel := factor(clinvarLabel, levels=c('Pathogenic/Likely pathogenic', 'Uncertain significance', 'Benign/Likely benign'))]
```

```{r, clinvar-mosaic-plot, fig.height=4, fig.width=6}
contrasts.oi <- unique(summary.dt$contrast)

summary.dt <- summary.dt[order(sigLabel)]
summary.dt[, colWidth := ifelse(sigLabel == 'no', 0.9, 0.5)] # just set width arbitrarily per group

lapply(contrasts.oi, function(x){
  
  g <- ggplot(summary.dt[contrast == x,], aes(x=sigLabel, width=colWidth, y=count, fill=clinvarLabel)) +
  geom_col(color='black',  position='fill') +
  #geom_label(data=summary.dt[contrast =='Alpelisib_22_vs_DMSO_22',], aes(y=ylab), color='black', fill='white', size=4) +
  geom_label(data=summary.dt[contrast == x,.(total=sum(count)), by=sigLabel], aes(y=1+.05, x=sigLabel, label=paste0('N=',as.character(total))), inherit.aes = F) +
  #facet_grid(~gene, scales = "free_x", space = "free_x")
  scale_fill_manual(values=c('Benign/Likely benign'=clinvar.pal[2], 'Uncertain significance'=clinvar.pal[5], 'Pathogenic/Likely pathogenic'=clinvar.pal[8])) + 
  labs(title=x, y='Proportion', x='|LFC| > 1 & pvalue < 0.005') +
  theme_classic() +
  guides(fill=guide_legend(title='Clinvar classification'))
  
  g
  #BackupAsPDF(g, paste0('clinvar_mosaic/',x,'.mosaic'))
})
```
Run the fishers exact test on this, see if we are enriched for pathogenic mutants in any of the groups
Simplify the clinvar label to see which of these are significantly different
```{r, fig.width=10, fig.height=6}
clinvar.expand[, clinvarBinary := 'Other']
clinvar.expand[grepl('pathogenic',ignore.case = T, clinvarLabel), clinvarBinary := 'Pathogenic/Likely pathogenic']

new.fisher.results <- clinvar.expand[ , 
  if (.N > 0L) {
    cont <- table(sigLabel, clinvarBinary)
    if (all(is.finite(cont)) && all(cont >= 0) && nrow(cont) > 1 && ncol(cont) > 1) {
      fisher.p = fisher.test(cont, alternative = 'greater')$p.value
    } else {
      fisher.p = 1 # another option??
    }
  }, by = .(contrast, gene)]

new.fisher.results[, padj := p.adjust(V1, method='BH')]
fwrite(new.fisher.results, ScriptAndDatedFileName('fishersTestresults.clinvar.enrichment.csv'))

g <- ggplot(new.fisher.results, aes(x=gene, y=-log10(V1), fill=gene, alpha=ifelse(V1 < 0.05, 'yes','no'))) +
  geom_hline(yintercept = -log10(0.05), linetype=2) +
  geom_bar(stat='identity') +
  facet_wrap(~contrast) +
  labs('results of fishers exact test', y='-log10 pvalue') +
  scale_fill_manual(values=unname(gene.col)) +
  scale_alpha_manual(values=c('yes'=1, 'no'=0.5)) +
  theme_bw() +
  customTheme +
  guides(alpha=guide_legend('signficant?'))
g
BackupAsPDF(g, 'fishers.test.res.barplot')
```

Ok, now we just need to plot the lollipop plot for each geen and specific contrasts 
Need annotation file for each of these; look at old plots for this

```{r}
# heres PIK3CA from earlier annotation/paper
anno.dt <- data.table(gene=rep('PIK3CA',5),     
                      domain=c('PI3K-ABD','PI3K-RBD', 'C2 PI3K-type', 'PIK helical', 'PI3K/PI4K catalytic'),
                      start=c(16,187, 330,517,765),
                      end=c(105,289,487,694,1051))

#PTEN
anno.dt <- rbind(anno.dt, data.table(gene=rep("PTEN", 2),
                          domain=c('Phosphatase tensin-type', 'C2 tensin-type'),
                          start=c(14, 190),
                          end=c(185,350))
                          )
# MTOR
anno.dt <- rbind(anno.dt, data.table(gene=rep("MTOR", 3),
                          domain=c('FAT', 'PI3K/PI4K catalytic', 'FATC'),
                          start=c(1382, 2156, 2517),
                          end=c(1982, 2469, 2549))
                          )


# AKT1
anno.dt <- rbind(anno.dt, data.table(gene=rep('AKT1', 3),
                                     domain=c('PH', 'Protein kinase', 'AGC-kinase C-terminal'),
                                     start=c(5,150,409),
                                     end=c(108,408,480)
                                     ))

```
Lollipop plots; first plot by genetype
Use the expanded data.table for now,
```{r}
de.dt <- fread('~/Documents/projects/101224_RBabu_CRISPRe_PIK3CA/111824_plotsForKroganWeekly_data/2025_01_22_deseq.pwcomparisons.allAnno.csv')

gene.oi <- c('MTOR','PIK3CA', 'PTEN', 'AKT1')

de.dt[, c('numerator', 'denominator') := tstrsplit(contrast, '_vs_', keep=c(1,2))]
de.dt[, numerator := factor(numerator, levels=c('DMSO_7','DMSO_22', 'Paxalisib_7','Paxalisib_22', 'Alpelisib_7','Alpelisib_22'))]
de.dt[, numeratorTreatment := gsub('_[0-9]{1,2}', '', numerator)]
de.dt[, Gene := gene] # just to avoid reediting


selected.targets <- setDT(read_xlsx('./data/ronald_targets.xlsx', col_names =F))
setnames(selected.targets, new=c('guideName', 'sequence','status'))

de.dt <- merge(de.dt, selected.targets[,.(guideName,sequence, status)], by.x=c('guide'), by.y=c('sequence'), all.x=T)

# need to tidy clinvar label or will throw errors if not in the supplied dt
de.dt[,clinvar_cs := clinicalSignificance]
de.dt[clinvar_cs == '', clinvar_cs := 'no annotation']

clinvar.severity <- data.table(Consequence =c(
                        'Pathogenic',
                        'Likely pathogenic',
                        'Pathogenic/Likely pathogenic',
                        'Conflicting classifications of pathogenicity',
                        'Uncertain significance',
                        'Likely benign',
                        "Benign/Likely benign",
                        'Benign',
                        'not provided',
                        'no annotation'))

de.dt[, clinvar_mostSevere := getMostSevereVariant(clinvar_cs, rankingDT=clinvar.severity), by=.I]
de.dt[, clinvar_mostCommon := getMostCommonVariant(clinvar_cs, rankingDT=clinvar.severity), by=.I]


de.dt[, clinvar_mostSevere := factor(clinvar_mostSevere, levels=c("no annotation", "not provided", "Benign", "Benign/Likely benign", "Likely benign", 
                                                        "Uncertain significance", "Conflicting classifications of pathogenicity",
                                                        "Likely pathogenic", "Pathogenic/Likely pathogenic", "Pathogenic"))]

de.dt[, clinvar_mostCommon := factor(clinvar_mostCommon, levels=c("no annotation", "not provided", "Benign", "Benign/Likely benign", "Likely benign", 
                                                        "Uncertain significance", "Conflicting classifications of pathogenicity",
                                                        "Likely pathogenic", "Pathogenic/Likely pathogenic", "Pathogenic"))]
```


```{r}
#fwrite(de.dt, ScriptAndDatedFileName('deseq2.pwcomparisons.lollipopToplot.csv.gz'))
de.dt <- fread('~/Documents/projects/101224_RBabu_CRISPRe_PIK3CA/050225_kroganWeeklyPlots_data/2025_05_05_deseq2.pwcomparisons.lollipopToplot.csv.gz')
expand.dt <- fread('~/Documents/projects/101224_RBabu_CRISPRe_PIK3CA/050225_kroganWeeklyPlots_data/2025_05_05_expanded.deseq.pwcomparisons.alphamissenseAnno.csv.gz')

#fwrite(anno.dt, ScriptAndDatedFileName('uniport.domains.csv.gz'))
anno.dt <- fread('~/Documents/projects/101224_RBabu_CRISPRe_PIK3CA/050225_kroganWeeklyPlots_data/2025_05_05_uniport.domains.csv.gz')
setkey(anno.dt, gene, start,end)
```
merge the annotation datatable to the de.dt using `foverlaps`

```{r}
# is this neccassary? can also specify by.x by.y whih is more explicit....
setkey(anno.dt, gene, start,end)

# add the same columns to the large LFC table
de.dt[, start := first_peptide_pos]
de.dt[, end := first_peptide_pos]

setkey(de.dt, gene, start,end)

anno.dt %>% head()
de.dt %>% head()

#using start and end keys for the PIK3CA lookup table
# by default, NA will be added to columns with no matching row
# for now dont expand multi rows
# maybe write a function to get the position. of the most severe variant
de.dt <- foverlaps(de.dt[!gene %in% c('INTERGENIC', 'NON-TARGETING') & first_peptide_pos != '',], anno.dt, type="any", nomatch=NA, mult='all')

de.dt
de.dt[is.na(domain), domain := 'na']
```
Take a single plot and generate an example
Look at the newscalefill package to inlcude multiple color scale fill 
```{r lollipop, fig.height=9, fig.width=14}

p <- ggplot(subdt, aes(x=first_peptide_pos, y=log2FoldChange, shape=editor, fill=lfcSig, label=pep_mutant_id, alpha=lfcSig)) +
    geom_segment(aes(x=first_peptide_pos, xend=first_peptide_pos, y=0, yend=log2FoldChange), color="grey") +
    geom_point(aes(size=-log10(pvalue)), show.legend = F) +
    geom_hline(yintercept=c(-1,1), alpha=0.6, linetype=3) +
    geom_hline(yintercept=c(-1.5,1.5), alpha=0.6, linetype=2) +
    #geom_hline(yintercept=0,size=4, aes(color=COLOR)) +
    #geom_link(aes(x = 0, xend = max(first_peptide_pos), y = -0.5, yend = 0.5, colour = COLOR), lineend = "round") +
    scale_x_continuous(n.breaks = 20, expand=c(0,0),limits=c(0,NA)) +
    facet_wrap(~gene, ncol=1, scales='free_x') +
    #labs(subtitle=x, x='amino acid residue', y='Log2FoldChange') +
    ggrepel::geom_text_repel(data=subdt[lfcSig != 'not' & !grepl('silent', most_common_mutant),], 
                             color='black',
                            # aes(color=screening),
                             show.legend = F, segment.alpha=0.5, segment.linetype=2, max.overlaps = 8, size=1.75) +
    # now label the selected guides
    #ggrepel::geom_text_repel(data=subdt[lfcSig != 'not' & gene %in% c("AKT1","MTOR","PIK3CA","PTEN") & !grepl('silent', most_common_mutant) & id %in% guidesOI,], 
    #                         show.legend = F, segment.alpha=0.5, segment.linetype=2, max.overlaps = 8, size=2, color='maroon') +   
    #scale_color_manual(values=c('selected for screening'='maroon', 'ns'='black')) +
    scale_fill_manual(values=c('not'='grey70', 'up'=redBlue[2], 'down'=redBlue[4]), name='abs(LFC) > 1') +
    scale_alpha_manual(values = c('up' = 1, 'down'=1, 'not' =  0.2)) +
    scale_size_continuous(breaks=c(0,1,3,4), range=c(1.5,6)) +
    #scale_shape_manual(values = 21:22) +
    scale_shape_manual(values = c(21,22)) + # to use boht color and fill
    theme_classic() +
   # customTheme +
    theme(plot.title = element_text(size = 20, face = "bold")) 
  
  p <- p +
    new_scale_fill() +
    #geom_path(data = line.data, inherit.aes = FALSE, aes(x = first_peptide_pos, y = y, color = domain, size=4)) +
    geom_gene_arrow(aes(xmin=start, xmax=end, y=0, fill=domain), inherit.aes = FALSE,
                    color='black', arrowhead_width = grid::unit(0, "mm"), arrowhead_height = grid::unit(0, "mm")) + 
    scale_fill_manual(values=c(domain.pal)) +
    guides(alpha='none',
           fill = guide_legend(override.aes = list(shape =21) ) ,
           size = 'none',
           color = guide_legend('Uniprot domain'))
    
  #BackupAsPDF(p, paste0('test','.tileplot'))
p
```
Loop throught the different contrasts and plot this

```{r lollipop, fig.height=9, fig.width=14}

lapply(unique(de.dt$contrast), function(x){
  
  subdt <- de.dt[contrast == x & !gene %in% c('INTERGENIC', 'NON-TARGETING'),]
  subdt[, lfcSig := 'not']
  subdt[abs(log2FoldChange) > 1.5, lfcSig := ifelse(log2FoldChange > 0, 'up', 'down')]
  subdt[, screening := ifelse(id %in% guidesOI, 'selected for screening', 'ns')]
  
  # loop over the different contrasts
  p <- ggplot(subdt, aes(x=first_peptide_pos, y=log2FoldChange, shape=editor, fill=lfcSig, label=pep_mutant_id, alpha=lfcSig)) +
    geom_segment(aes(x=first_peptide_pos, xend=first_peptide_pos, y=0, yend=log2FoldChange), color="grey") +
    geom_point(aes(size=-log10(pvalue)), show.legend = F) +
    geom_hline(yintercept=c(-1,1), alpha=0.6, linetype=3) +
    geom_hline(yintercept=c(-1.5,1.5), alpha=0.6, linetype=2) +
    scale_x_continuous(n.breaks = 20, expand=c(0,0),limits=c(0,NA)) +
    facet_wrap(~gene, ncol=1, scales='free_x') +
    labs(title=x, x='amino acid residue', y='Log2FoldChange') +
    ggrepel::geom_text_repel(data=subdt[lfcSig != 'not' & !grepl('silent', most_common_mutant),], 
                             aes(color=ifelse(id %in% guidesOI, 'screening target', 'ns'), size=ifelse(id %in% guidesOI, 2.25, 1.9)),
                             show.legend = F, segment.alpha=0.5, segment.linetype=2, max.overlaps = 4) +
    scale_color_manual(values=c('ns'='black', 'screening target'='maroon')) +                           
    scale_fill_manual(values=c('not'='grey70', 'up'=redBlue[2], 'down'=redBlue[4]), name='abs(LFC) > 1') +
    scale_alpha_manual(values = c('up' = 1, 'down'=1, 'not' =  0.2)) +
    scale_shape_manual(values = c(21,22)) + # to use boht color and fill
    theme_classic() +
    theme(plot.title = element_text(size = 20, face = "bold")) 
  
  # update plot with annotations
  p <- p +
    new_scale_fill() + # clean fill aes to include domain info
    geom_gene_arrow(aes(xmin=start, xmax=end, y=0, fill=domain), inherit.aes = FALSE,
                    color='black', arrowhead_width = grid::unit(0, "mm"), arrowhead_height = grid::unit(0, "mm")) + 
    scale_fill_manual(values=domain.pal) +
    guides(alpha='none',
           fill = guide_legend(override.aes = list(shape =21)) ,
           size = 'none')
    
  #BackupAsPDF(p, paste0('lollipop_contrasts/', x, '.lolipop.plot'))
  p
})
```

Ok, now we wan to plot this for a couple of contrasts and a couple of genes,  nothing too major

```{r, lollipop-genes, fig.height=8, fig.width=14}

genes.oi <- c('MTOR','AKT1', 'PTEN', 'PIK3CA')
contrasts.oi <- c("Alpelisib_22_vs_LibraryRep_0", "Paxalisib_22_vs_LibraryRep_0", "DMSO_22_vs_LibraryRep_0")


lapply(unique(genes.oi), function(x){
  
  subdt <- de.dt[contrast %in% contrasts.oi & gene == x,]
  subdt[, lfcSig := 'not']
  subdt[abs(log2FoldChange) > 1.5, lfcSig := ifelse(log2FoldChange > 0, 'up', 'down')]
  subdt[, screening := ifelse(id %in% guidesOI, 'selected for screening', 'ns')]
  
  subdt[, contrast := factor(contrast, levels=c("DMSO_22_vs_LibraryRep_0", "Alpelisib_22_vs_LibraryRep_0", "Paxalisib_22_vs_LibraryRep_0"))]
  
  # loop over the different contrasts
  p <- ggplot(subdt, aes(x=first_peptide_pos, y=log2FoldChange, shape=editor, fill=lfcSig, label=pep_mutant_id, alpha=lfcSig)) +
    geom_segment(aes(x=first_peptide_pos, xend=first_peptide_pos, y=0, yend=log2FoldChange), color="grey") +
    geom_point(aes(size=-log10(pvalue)), show.legend = F) +
    geom_hline(yintercept=c(-1,1), alpha=0.6, linetype=3) +
    geom_hline(yintercept=c(-1.5,1.5), alpha=0.6, linetype=2) +
    scale_x_continuous(n.breaks = 20, expand=c(0,0),limits=c(0,NA)) +
    facet_wrap(~contrast, ncol=1, scales='free_x') +
    labs(title=x, x='amino acid residue', y='Log2FoldChange') +
    ggrepel::geom_text_repel(data=subdt[lfcSig != 'not' & !grepl('silent', most_common_mutant),], 
                             aes(color=ifelse(id %in% guidesOI, 'screening target', 'ns'), size=ifelse(id %in% guidesOI, 2.25, 1.9)),
                             show.legend = F, segment.alpha=0.5, segment.linetype=2, max.overlaps = 4) +
    scale_color_manual(values=c('ns'='black', 'screening target'='maroon')) +                           
    scale_fill_manual(values=c('not'='grey70', 'up'=redBlue[2], 'down'=redBlue[4]), name='abs(LFC) > 1') +
    scale_alpha_manual(values = c('up' = 1, 'down'=1, 'not' =  0.2)) +
    scale_shape_manual(values = c(21,22)) + # to use boht color and fill
    theme_classic() +
    theme(plot.title = element_text(size = 20, face = "bold")) 
  
  # update plot with annotations
  p <- p +
    new_scale_fill() + # clean fill aes to include domain info
    geom_gene_arrow(aes(xmin=start, xmax=end, y=0, fill=domain), inherit.aes = FALSE,
                    color='black', arrowhead_width = grid::unit(0, "mm"), arrowhead_height = grid::unit(0, "mm")) + 
    scale_fill_manual(values=domain.pal) +
    guides(alpha='none',
           fill = guide_legend(override.aes = list(shape =21)) ,
           size = 'none')
    
  BackupAsPDF(p, paste0('lollipop_genes/', x, '.lolipop.plot'))
  p
})
```
Last plot; ridgeplots for alpha missense to see how they move over time for each of the genes
Do all genes

```{r}
expand.dt <- fread('~/Documents/projects/101224_RBabu_CRISPRe_PIK3CA/050225_kroganWeeklyPlots_data/2025_05_05_expanded.deseq.pwcomparisons.alphamissenseAnno.csv.gz')

contrasts.oi <- c("Alpelisib_22_vs_LibraryRep_0", "Paxalisib_22_vs_LibraryRep_0", "DMSO_22_vs_LibraryRep_0",
                  "Alpelisib_7_vs_LibraryRep_0", "Paxalisib_7_vs_LibraryRep_0", "DMSO_7_vs_LibraryRep_0"
                  )
test.dt <- expand.dt[gene == 'PTEN' & contrast %in% contrasts.oi]

test.dt[, numeratorTreatment := factor(numeratorTreatment, levels=c('DMSO', 'Paxalisib', 'Alpelisib'))]
test.dt[, numeratorTimepoint := as.factor(numeratorTimepoint)]


g <- ggplot(test.dt[editor=='abe8e'], aes(x=log2FoldChange, y=numeratorTreatment, fill=numeratorTimepoint, alpha=numeratorTimepoint,  height = after_stat(density))) +
  ggridges::geom_density_ridges( 
  scale=1,
  jittered_points = TRUE,
  position = position_points_jitter(width = 0.05, height = 0),
  point_shape = '|', point_size = 3, point_alpha = 1, alpha = 0.7) +
  labs(title='test', y='treatment') +
  geom_vline(xintercept=c(-1,1), linetype=2, alpha=0.6) +
  scale_alpha_manual(values=c('7'=0.95,'22'=0.7)) +
  scale_fill_manual(values=c( "skyblue2","#FB9A99")) +
 # scale_color_manual(values=c( "skyblue2","#FB9A99")) +
  theme_classic()
g
BackupAsPDF(g, 'test.ridgeplot.rug')
```



```{r, ridgeplot-bygene, fig.width=10, fig.height=5}

lapply(unique(genes.oi), function(x){
  
  subdt <- expand.dt[contrast %in% contrasts.oi & gene == x,]
  #subdt[, contrast := factor(contrast, levels=c("DMSO_22_vs_LibraryRep_0", "Alpelisib_22_vs_LibraryRep_0", "Paxalisib_22_vs_LibraryRep_0", 
  #                                              "DMSO_7_vs_LibraryRep_0", "Alpelisib_7_vs_LibraryRep_0", "Paxalisib_7_vs_LibraryRep_0"))]
  
  subdt[, numeratorTreatment := factor(numeratorTreatment, levels=c('DMSO', 'Paxalisib', 'Alpelisib'))]
  subdt[, numeratorTimepoint := factor(numeratorTimepoint, levels = c("7", "22"))]
  subdt[, am_class := factor(am_class, levels=c('benign', 'ambiguous', 'pathogenic'))]
  # loop over the different contrasts
  g <- ggplot(subdt[editor=='abe8e'], aes(x=log2FoldChange, y=numeratorTreatment, fill=am_class, alpha=numeratorTimepoint,  height = after_stat(density))) +
    ggridges::geom_density_ridges( 
    aes(linetype=numeratorTimepoint),
    scale=1,
    jittered_points = FALSE,
    position = position_points_jitter(width = 0.05, height = 0),
    point_shape = '|', point_size = 3, point_alpha = 1) +
    labs(title='LFC distributions of AlphaMissense-annotated mutatations', subtitle=x, y='treatment') +
    scale_y_discrete(expand = c(0, 0)) +
    geom_vline(xintercept=c(-1,1), linetype=2, alpha=0.6) +
    scale_alpha_manual(values=c('7'=0.9,'22'=0.6)) +
    scale_fill_manual(values=c(pathogenic=col.pal[2], ambiguous='grey70', benign='cornflowerblue')) +
    #scale_fill_manual(values=c( "skyblue2","grey70","#FB9A99")) +
    facet_grid(~am_class) +
    # scale_color_manual(values=c( "skyblue2","#FB9A99")) +
    theme_classic()
  g
  BackupAsPDF(g, paste0(x, '.ridgeplot.alphamissense.rug'))
})

```
```{r, ridgeplot-bygene, fig.width=10, fig.height=5}

lapply(unique(genes.oi), function(x){
  
  subdt <- expand.dt[contrast %in% contrasts.oi & gene == x,]
  subdt[, numeratorTreatment := factor(numeratorTreatment, levels=c('DMSO', 'Paxalisib', 'Alpelisib'))]
  subdt[, numeratorTimepoint := factor(numeratorTimepoint, levels = c("7", "22"))]
  subdt[, am_class := factor(am_class, levels=c('benign', 'ambiguous', 'pathogenic'))]
  # loop over the different contrasts
 
   g <- ggplot(subdt[editor=='abe8e'], aes(x=log2FoldChange, y=numeratorTreatment, fill=numeratorTimepoint, alpha=numeratorTimepoint, height = after_stat(density))) +
    ggridges::geom_density_ridges( 
    aes(linetype=numeratorTimepoint, point_colour = numeratorTimepoint),
    scale=1,
    jittered_points = TRUE,
    position = position_points_jitter(width = 0.05, height = 0),
    point_shape = '|', point_size = 3, point_alpha = 0.6) +
    labs(title='LFC distributions of AlphaMissense-annotated mutatations', subtitle=x, y='treatment') +
    scale_y_discrete(expand = c(0, 0)) +
    geom_vline(xintercept=c(-1,1), linetype=2, alpha=0.6) +
    scale_alpha_manual(values=c('7'=0.9,'22'=0.6)) +
    scale_fill_manual(values=c("#A6D854", "#FFD92F")) +
    scale_color_manual(values=c("#A6D854", "#FFD92F")) +
    scale_discrete_manual("point_color", values=c("#A6D854", "#FFD92F")) +
    facet_grid(~am_class) +
    theme_classic()
  g
  BackupAsPDF(g, paste0(x, '.ridgeplot.alphamissense.adjColor.rug'))
})
```

Double check the fishers exact test results; ensure w

```{r}

```

**not used**
```{r}
expand.dt
lapply(genes.oi, function(x){

    sub.dt <- expand.dt[gene == 'MTOR' & contrast %in% contrasts.oi,]
    # take tidy the naming convention
  
    
     g <- ggplot(sub.dt, aes(x=log2FoldChange, y=am_class)) + #,  height = after_stat(density))) +
      geom_density(alpha=0.7) +
      xlab('Log2FC') +
      ylab('Mutation Consequence') +
      geom_vline(xintercept=c(-1,1), linetype=2, alpha=0.6) +
     # scale_x_continuous(labels=seq(-4,4,1)) +
      #ggtitle(paste(y, x, sep=' ')) +
      #scale_fill_manual(values=col.pal) +
     # scale_fill_viridis_d(option=2) +
     facet_grid(am_class~numeratorTreatment) +
      theme_bw() +
      theme(strip.text.y=element_blank(),
            axis.text.y = element_blank()
            ) +
      guides(fill= guide_legend('Category'))
    g
    #BackupAsPDF(g, paste0('ridgeplots/mutant_anno/bygene_mostSevereMutant/',x,'.', y, '.LFC.Distributions.mutantConsequence.ridgeplot'), dimensions = c(10,8))
})

```

```{r}

lapply(genes.oi, function(x){

    sub.dt <- de.dt[gene %in% c('MTOR','PIK3CA', 'PTEN', 'AKT1') & contrast %in% contrasts.oi,]
    # take tidy the naming convention
    
    lapply(unique(sub.dt$gene), function(y){
      
     g <- ggplot(sub.dt[gene == y & !is.na(most_severe_mutant) & most_severe_mutant != 'nonsense',], aes(x=log2FoldChange, fill=most_severe_mutant)) + #,  height = after_stat(density))) +
      geom_density(alpha=0.7) +
      xlab('Log2FC') +
      ylab('Mutation Consequence') +
      geom_vline(xintercept=c(-1,1), linetype=2, alpha=0.6) +
     # scale_x_continuous(labels=seq(-4,4,1)) +
      ggtitle(paste(y, x, sep=' ')) +
      #scale_fill_manual(values=col.pal) +
      scale_fill_viridis_d(option=2) +
      facet_grid(most_severe_mutant~editor) +
      theme_ridges() +
      theme(strip.text.y=element_blank(),
            axis.text.y = element_blank()
            ) +
      guides(fill= guide_legend('Category'))
    g
    #BackupAsPDF(g, paste0('ridgeplots/mutant_anno/bygene_mostSevereMutant/',x,'.', y, '.LFC.Distributions.mutantConsequence.ridgeplot'), dimensions = c(10,8))
    })
})
```


```{r}
lapply(unique(de.dt$contrast), function(x){

    sub.dt <- de.dt[gene %in% c('MTOR','PIK3CA', 'PTEN', 'AKT1') & contrast == x,]
    # take tidy the naming convention
    
    lapply(unique(sub.dt$gene), function(y){
      
     g <- ggplot(sub.dt[gene == y & most_common_mutant != 'nonsense',], aes(x=log2FoldChange, fill=most_common_mutant)) + #,  height = after_stat(density))) +
      geom_density(alpha=0.7) +
      xlab('Log2FC') +
      ylab('Mutation Consequence') +
      geom_vline(xintercept=c(-1,1), linetype=2, alpha=0.6) +
     # scale_x_continuous(labels=seq(-4,4,1)) +
      #ggtitle(paste('PIK3CA', title, sep=' ')) +
     # ggtitle(plotTitle) +
      ggtitle(paste('PIK3CA', x, sep=' ')) +
      scale_fill_viridis_d(option=2) +
      facet_grid(most_common_mutant~editor) +
      theme_ridges() +
      theme(strip.text.y=element_blank(),
            axis.text.y = element_blank()
            ) +
      guides(fill= guide_legend('Category'))
    g
   # BackupAsPDF(g, paste0('ridgeplots/clinvar_anno/bygene_mostSevereMutant/',x,'.', y, '.LFC.Distributions.mutantConsequence.ridgeplot'), dimensions = c(10,8))
    })
})
```



