---
title: "091824_dataPreparationforJyoti"
author: "Martin Gordon"
date: "2024-09-18"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

Goals

i) Produce plot of relative abundances of viral and host mapped reads per cell-line
ii) Produce a CPM plot: Counts per million (CPM) mapped reads are the number of raw reads mapped to a transcript, scaled by the number of sequencing reads in your sample, multiplied by a million
  Is this metric suitable? We are only counting hits to leader +1 position, so I think its ok (I dont think we need to consider gene length as only counting 5' end of transcript)
iii) Produce normalized counts for both the mouse and human RNAseq datasets

```{r}
library(DESeq2)
library(data.table)
library(magrittr)
library(ggplot2)
library(stringr)
library(circlize)
library(ComplexHeatmap)
library(ggrepel)
library(readxl)
library(usedist) #package for working with distances
library(ggh4x) # additional functionality for ggplot2 obj; eg facet_grid allow x/y axis to vary
library(scales)
library(patchwork)
library(RColorBrewer)
library(cluster) # pam clustering of genes
library(eulerr) # eulerr plot 
library(ggvenn)
library(viridis)
library(readxl)
library(viridis)

library(biomaRt)

# RNAseq DE functions
library(tximport)
library(DESeq2)
library(IHW)

source("../../utils/bp_utils/ManageScriptData.R")
source ("../../utils/bp_utils/enrichmentTestFunctions.R")
source ("../../utils/bp_utils/UniprotIDMapping.R") #uniprot to gene id mapping

source("../../utils/mg_utils/r_utils/IDmapping.R")
source("../../utils/mg_utils/r_utils/plottingHelperFunctions.R")
source("../../utils/mg_utils/r_utils/HelperFunctions.R")
source("../../utils/mg_utils/r_utils/differentialTesting.R")
source("/Users/martingordon/Documents/utils//bp_utils/LinearModels.R")

# function to cluster w NA values
clusterWNA <-  function(mat, na.value=0,...){
  
  mat[is.na(mat)] <- na.value
  return(hclust(dist(mat)))
}

redbluColpal <- c('#D01B1B', '#FF4242', '#FFFFFF', '#95D2EC','#47abd8')
```

read in the dds objects for bat and human and generate the normalized count matrices
```{r}
hu.dds <- readRDS('./)
```

