---
title: "AllSearchesComparison"
author: "Martin Gordon"
date: "2024-02-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(magrittr)
library(data.table)
library(circlize)
library(ComplexHeatmap)
library(circlize)
library(ggplot2)
library(stringr)
library(randomcoloR)
library(MSstats)
library(ggrepel)
library(readxl)
library(nichenetr)
library(patchwork) # combine plots
library(viridis)
library(ggvenn)


source("../../utils/bp_utils/MSstats_Helper_Functions.R")
source("../../utils/bp_utils/ManageScriptData.R")
source ("../../utils/bp_utils/MSstats_V4_Functions.R")
source ("../../utils/bp_utils/UniprotIDMapping.R") #uniprot to gene id mapping
source ("../../utils/bp_utils/enrichmentTestFunctions.R")

# function to cluster w NA values
clusterWNA <-  function(mat, na.value=0){
  
  mat[is.na(mat)] <- na.value
  return(hclust(dist(mat)))
}
```


```{r}
spec.list <-  list(dda_bg = fread('./data/DDA Lib.tsv'),
                   dda_noBg = fread('./data/DDA Lib No Background.tsv'),
                   dual_bg = fread('./data/Dual Lib .tsv'),
                   dual_noBg = fread('./data/2024.02.21 - Mac1 AB - Dual No Bg_Report_MSStats Format (Normal).tsv'),
                   allSamples = fread('../020524_AVIDD_ABproteomics/data/AB_Report_MSstats_Format.tsv'))

# sample ids 
keys <- fread('./data/keys.txt')
keys
keys[,.N, .(Condition,RawFile)]
# contrasts
contrasts.txt <- fread('./data/contrasts.txt', sep='-', header=F)
```

```{r}
# the dual DIA and DDA is a much bigger library; more things detected, but perhaps also an issue
lapply(spec.list, dim)
lapply(spec.list, str)
```

combine with sample metadata
```{r}
keys$RawFile %>%  unique()
lapply(spec.dt.list, function(x){
  x$Run %>%  unique()
})

spec.dt.list <- lapply(spec.list, function(X){
  
  spec.dt <- merge(x=X[,!c("Condition", "BioReplicate")], y=keys[,.(Condition, BioReplicate, RawFile, Fraction,IsotopeLabelType)], by.x='Run', by.y='RawFile')
  return(spec.dt)
})
```

Read in the MOUSE ID mapping file 

```{r}
# lets map the homologs to the mss output and write the complete and tidied data to file 
mh_mappings <- fread('~/Documents/utils/mg_utils/data/mouseHumanIDConversion.txt')

# gene sets OI
piGenes <- read_xlsx('./docs/1-s2.0-S0092867423009157-mmc4.xlsx', col_names = T, sheet=7) %>% 
  setDT(.) %>% 
  .[category == 'Proinflammatory'] %>% 
  .[,unique(gene)]

isGenes <- fread('./docs/ISGs.txt', header=F) %>% 
  .[,V1]
```