---
title: "PWComparisonsRedo"
author: "Martin Gordon"
date: "2024-04-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Regenerate the PW comparisons of the different groups with the second set of controls (drop contorl group1)
## Using DMSO ctrl group 2 as  


Questions:
Should we filter out guides rather than just removing 0 counts? What about removing sgRNA with 0 count in controls? before DE testing?
Many of our variants are of moderate impact (missense variants)... would we not expect more synomonous? Need to look into this... also v different breakdowns for the different groups.. no real difference in variant distribution by significance..
Redo the dotplot with protein coordinates

Why are we losing so many variants? Output by vep was approx 12k guide RNA.. losing ~ 1/4 with filtering..investigate
  - Have missing sgRNAs.. look up VEP files and recover these sgRNAs and add back in.. how many of these are the intergenic/non-targeting sgRNA? ~ 1k so still many more missing
  - Rerun VEP on CL... is it failing with incorrect ref??
  - Redo basic LFC and just see if we recover more downreg before trying mageck
  
  
I think the key question is: Is the data poor quality or are the annotations wrong? If the annotations are wrong, will plotting the position help? Maybe also merge with the sgRNA dt seperately, and just plot by gene and posiiton to see if there is a trend that is mimiced by the annotated data
  
```{r}
library(magrittr)
library(data.table)
library(circlize)
library(ComplexHeatmap)
library(ggplot2)
library(stringr)
library(randomcoloR)
library(ggrepel)
library(ggridges) # ridgeline plots
library(eulerr) # for venn diagram plots
library(ggvenn)

source("../../utils/bp_utils/MSstats_Helper_Functions.R")
source("../../utils/bp_utils/ManageScriptData.R")
source ("../../utils/bp_utils/enrichmentTestFunctions.R")
source ("../../utils/mg_utils/r_utils/IDmapping.R")
source ("../../utils/mg_utils/r_utils/plottingHelperFunctions.R")

# function to cluster w NA values
clusterWNA <-  function(mat, na.value=0,...){
  
  mat[is.na(mat)] <- na.value
  return(hclust(dist(mat)))
}
```

First thing to do is to regenerate the count matrices

read in the count matrix taking the logNormalised counts 

```{r}
counts.dt <- fread('./032824_MAGECK_DE_firstPass_data/2024_03_28_sampleCounts.csv')
```

```{r}
norm.mat.dt <- fread("/Users/martingordon/Documents/projects/032124_RBabu_CRISPRMutantScreen/032824_MAGECK_DE_firstPass_data/2024_04_02_normalizedCounts.matrix.csv")

# first thing we need to do is write to each file a subset of the mat of interest
selectCols <- function(dt, colPatterns=c('sgRNA', 'Gene', 'Abe8eDMSO','Abe8eSerum')){
  cols.oi <- grep(paste(colPatterns, collapse = '|'), colnames(dt), value = T)
  
  print(cols.oi)
  dt[, ..cols.oi]
}

# first grp is fitness analysis after base editing
cols.oi <- list(c('sgRNA', 'Gene', 'Bemax_Lib', 'BemaxDMSO_CTRL2'),
                c('sgRNA', 'Gene', 'Bemax_Lib', 'BemaxDMSO_CTRL2'),
                c('sgRNA', 'Gene', 'Abe8e_Lib', 'Abe8eDMSO_CTRL2'),
                c('sgRNA', 'Gene', 'Abe8e_Lib', 'Abe8eDMSO_CTRL2'),
                # drug sensitivity analysis
                #bemax set
                c('sgRNA', 'Gene', 'BemaxDMSO_CTRL2', 'Bemax_TORIN'),
                c('sgRNA', 'Gene', 'BemaxDMSO_CTRL2', 'Bemax_BAFA1'),
                c('sgRNA', 'Gene', 'BemaxDMSO_CTRL2', 'BemaxSerum_STARVE'),
                c('sgRNA', 'Gene', 'BemaxDMSO_CTRL2', 'Bemax_GPN'),
                c('sgRNA', 'Gene', 'BemaxDMSO_CTRL3', 'Bemax_DOXO'),
                #abe8e set
                c('sgRNA', 'Gene', 'Abe8eDMSO_CTRL2', 'Abe8e_TORIN'),
                c('sgRNA', 'Gene', 'Abe8eDMSO_CTRL2', 'Abe8e_BAFA1'),
                c('sgRNA', 'Gene', 'Abe8eDMSO_CTRL2', 'Abe8eSerum_STARVE'),
                c('sgRNA', 'Gene', 'Abe8eDMSO_CTRL2', 'Abe8e_GPN'),
                c('sgRNA', 'Gene', 'Abe8eDMSO_CTRL3', 'Abe8e_DOXO'))
#

comparisons.list <- lapply(cols.oi, function(x){
  
   selectCols(norm.mat.dt, colPatterns = x)
})
names(comparisons.list) <- lapply(cols.oi, function(x){ paste(x[c(3:4)], collapse='-')  })

names(comparisons.list)

# wrrite the datatables to file
lapply(seq_along(comparisons.list), function(x,n,i){
  
  fwrite(x[[i]], paste0('./output/matrices/', n[[i]], '.160424.norm.count.tsv'), sep='\t')
  
}, x=comparisons.list, n=names(comparisons.list))
```

160424
----
Redo the comparisons using the CTRL2 group

Read in the MAGECK processed data
- read in the sgRNA data and ii) the gene level data

```{r}
de.sgRNA.meta <- dir(path='./output/mageck_test', pattern='*.sgrna_summary.txt', full.names = T, recursive = F) %>% 
  as.data.table()
de.gene.meta <- dir(path='./output/mageck_test', pattern='*.gene_summary.txt', full.names = T, recursive =F) %>% 
  as.data.table()

de.sgRNA.meta[, contrast := sub('.sgrna_summary.txt', '', sub('./output/mageck_test/', '', .)) ]
de.sgRNA.meta[, c('numerator', 'denominator') :=  tstrsplit(contrast, '-', keep=c(1,2))]

de.gene.meta[, contrast := sub('.sgrna_summary.txt', '', sub('./output/mageck_test/', '', .)) ]
de.gene.meta[, c('numerator', 'denominator') :=  tstrsplit(contrast, '-', keep=c(1,2))]
```
read in the output results and bind the DTs together

```{r}
de.sgRNA.list <- lapply(de.sgRNA.meta$., fread)
names(de.sgRNA.list) <- de.sgRNA.meta$contrast

# ok the results look good, bind and lets visualise hits
sgRNA.dt <- rbindlist(de.sgRNA.list, idcol = 'contrast')

sgRNA.dt[, c('Treatment', 'Control') := tstrsplit(contrast, '-', keep=c(1,2))]


sgRNA.dt <- sgRNA.dt[, c("contrast","sgrna","Gene","Control", "Treatment", "control_count","treatment_count","control_mean","treat_mean","LFC","control_var",
             "adj_var","score","p.low","p.high","p.twosided","FDR","high_in_treatment")]


#fwrite(sgRNA.dt, ScriptAndDatedFileName('sgRNA.mageck.test.redoWDMSOCtrl2.csv'))

sgRNA.wide <- dcast(sgRNA.dt, sgrna+Gene~contrast, value.var = c('LFC', 'p.low', 'p.high', 'p.twosided', 'FDR', 'control_count', 'treatment_count'))
#fwrite(sgRNA.wide, ScriptAndDatedFileName('sgRNA.mageck.test.redoWDMSOCtrl2.wideformat.csv'))
```

Plots of the analysis
We set a stronger LFC threshold here (halfing/doubling) than previous
Considering variance, we ideally only want to see the strongest signals from the data.. can adjust back down if not finding anything of interest

```{r}
# add a sig title
sgRNA.dt[,sig := 'not']
sgRNA.dt[abs(LFC) >= 1 & FDR < 0.05, sig := ifelse(LFC > 0, 'up', 'down')]

sgRNA.dt[, cas9 := ifelse(grepl('Abe8e', contrast), 'Abe8e', 'Bemax')]
sgRNA.dt[, grp_comparison := paste0(sub('Abe8e[_]{0,1}|Bemax[_]{0,1}', '', Treatment), '-vs-', sub('Abe8e[_]{0,1}|Bemax[_]{0,1}', '', Control))]
```


```{r}
ggplot(sgRNA.dt, aes(x=-log10(FDR), y=score)) +
  geom_point()
```

First; bar plots of sig hits per contrast

```{r}
# breakdown by contrast
g <- ggplot(sgRNA.dt[,.N, by=.(sig,grp_comparison, cas9)][sig != 'not',], aes(x=cas9, y=N, fill=sig)) +
  geom_bar(stat='Identity') +
  facet_wrap(~grp_comparison) +
  ggtitle('Significant sgRNA (Log2FC >= 1 & p.adjust < 0.05)') +
  scale_fill_manual(values=c('up'='#990033', 'down'='#336699')) +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90))
g

BackupAsPDF(g, 'sgRNA.sigHits.breakdownbyContrast', dimensions=c(10,12))
```

```{r}
# breakdown by contrast and gene
g <- ggplot(sgRNA.dt[,.N, by=.(sig,grp_comparison,Gene, cas9)][sig != 'not',], aes(x=cas9, y=N, fill=sig)) +
  geom_bar(stat='Identity') +
  facet_grid(Gene~grp_comparison) +
  ggtitle('Significant sgRNA (Log2FC >= 1 & p.adjust < 0.05)') +
  scale_fill_manual(values=c('up'='#990033', 'down'='#336699')) +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90),
        strip.text.x = element_text(size=7))
g

BackupAsPDF(g, 'sgRNA.sigHits.breakdownbyGeneandContrast', dimensions=c(11,14))
```

Breakdown by gene

```{r}
# breakdown by  and gene
g <- ggplot(sgRNA.dt[,.N, by=.(sig,Gene, cas9)][sig != 'not',], aes(x=cas9, y=N, fill=sig)) +
  geom_bar(stat='Identity') +
  facet_wrap(~Gene) +
  ggtitle('Significant sgRNA (Log2FC >= 1 & p.adjust < 0.05)') +
  scale_fill_manual(values=c('up'='#990033', 'down'='#336699')) +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90))
g
BackupAsPDF(g, 'sgRNA.sigHits.breakdownbyGene', dimensions=c(11,14))
```

Rank plots 
--
Ranking the FC of the different sgRNA per contrast
```{r}
lapply(unique(sgRNA.dt$contrast), function(x){

  subdt <- sgRNA.dt[contrast == x,]
    
  g <- ggplot(data=subdt, aes(x=reorder(sgrna,LFC), y=LFC, label=sgrna, col=sig)) +
    geom_point(size=1, aes(col=ifelse(LFC > 1, 'up', 
                                      ifelse(LFC< -1,'down', 'not')))) +
    #geom_hline(aes(alpha=0.6), yintercept = 0, linetype='dotdash') +
    geom_label_repel(data=subdt[sig != 'not',], aes(label=sgrna),  segment.linetype=3, max.overlaps = 20,size=1.5) +
    ggtitle(paste0(x, ' rank plot')) +
    theme_classic() +
    xlab('Rank') +
    # here we add numeric vals to expand the scales
    scale_x_discrete(expand = expansion(add=500)) +
    scale_y_discrete(expand = expansion(add=1)) +
    scale_color_manual(values=c('up'='#990033', 'down'='#336699', 'not'='grey')) +
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          legend.position = 'none'
        )
  
 BackupAsPDF(g, paste0('rankplots/', x, '.LFC.rankplot'), dimensions = c(6.5,10))
})
```

Heatmap of sig sgRNA
# about 1/6 of the datset is DE.. should this be a consideration for our normalization approach?
```{r}
sigHits <-  sgRNA.dt[FDR < 0.05 & abs(LFC) >= 1, unique(sgrna)]
counts.dt[, Replicate := seq_len(.N), by=.(group, sgRNA, Gene)]
```

Plot heatmap
I think this heatmap is kind of interesting; looks like the control vs Library comparison and DOXO are just complete outliers in both groups but relative treatment similarity is quite similiar within groups.

```{r}
col.pal <-  randomcoloR::distinctColorPalette(k=length(unique(counts.dt$group)))

sg.mat <- dcast(sgRNA.dt[sgrna %in% sigHits,], sgrna~contrast, value.var = 'LFC') %>% 
  as.matrix(rownames='sgrna')

# create annotation
ann <- data.table(treatment=str_extract(colnames(sg.mat), 'BAFA1|Lib|DOXO|STARVE|TORIN|GPN'),
                  cas9=ifelse(grepl('Abe8e|Bemax', colnames(sg.mat)), str_extract(colnames(sg.mat), 'Abe8e|Bemax'), 'Ctrl'))

colours <- list('cas9' = c('Abe8e' = '#4477AA', 'Bemax'="#228833", 'Ctrl'="#CCBB44"),
                'treatment' =c('BAFA1' = col.pal[1], 'Lib'=col.pal[2], 'DOXO'=col.pal[3], 'CTRL1'=col.pal[4], 'CTRL2'=col.pal[5], 'CTRL3'=col.pal[6],
                           'STARVE' = col.pal[7], 'TORIN'=col.pal[8], 'GPN'=col.pal[9], 'plasmid_BE'=col.pal[10], 'gRNA_noBE'=col.pal[11]))

colAnn <- HeatmapAnnotation(df = ann, col = colours)

hm <- Heatmap(sg.mat, 
        cluster_rows = clusterWNA(sg.mat, method='pearson'),
        top_annotation = colAnn,
        row_title = paste0(nrow(sg.mat), ' sgRNA (Log2FC >= 1 & p.adj < 0.05)'),
        name='Log2(FC)',
        col=colorRamp2(breaks=c(-2,0,2), colors=c('blue', 'white', 'red')),
        column_names_gp = gpar(fontsize=6),
        show_row_names = F)
hm
BackupAsPDF(draw(hm, column_title='Significant sgRNA'), 'sigsgRNA.LFC.pearsonDist.Heatmap', dimensions=c(10,10))
```
Replot but drop the DMSO Ctrl vs Lib group
Need to check quality of the DOXO and serum starve samples

```{r}
sg.mat <- dcast(sgRNA.dt[sgrna %in% sigHits,], sgrna~contrast, value.var = 'LFC') %>% 
  as.matrix(rownames='sgrna')

submat <-  sg.mat[, grep('CTRL2[-].+', colnames(sg.mat), invert=T)]
# create annotation
ann <- data.table(treatment=str_extract(colnames(submat), 'BAFA1|DOXO|STARVE|TORIN|GPN'),
                  cas9=ifelse(grepl('Abe8e|Bemax', colnames(submat)), str_extract(colnames(submat), 'Abe8e|Bemax'), 'Ctrl'))

colours <- list('cas9' = c('Abe8e' = '#4477AA', 'Bemax'="#228833", 'Ctrl'="#CCBB44"),
                'treatment' =c('BAFA1' = col.pal[1], 'Lib'=col.pal[2], 'DOXO'=col.pal[3],
                           'STARVE' = col.pal[7], 'TORIN'=col.pal[8], 'GPN'=col.pal[9]))

colAnn <- HeatmapAnnotation(df = ann, col = colours)

hm <- Heatmap(submat, 
        cluster_rows = clusterWNA(submat, method='pearson'),
        top_annotation = colAnn,
        row_title = paste0(nrow(submat), ' sgRNA (Log2FC >= 1 & p.adj < 0.05)'),
        name='Log2(FC)',
        col=colorRamp2(breaks=c(-2,0,2), colors=c('blue', 'white', 'red')),
        column_names_gp = gpar(fontsize=6),
        show_row_names = F)
hm

BackupAsPDF(draw(hm, column_title='Significant sgRNA'), 'sigsgRNA.LFC.pearsonDist.noCtrlvsLib.Heatmap', dimensions=c(10,10))
```

Plot the expression heatmap of these sig hits
Drop CTRL1 group 

```{r}
norm.mat <-  dcast(counts.dt, sgRNA_Gene~sample, value.var='tmo.normCounts') %>% 
  as.matrix(rownames='sgRNA_Gene')

submat <- norm.mat[gsub('[_]TFEB|[_]COLN1|[_]TFE3|[_]PPP3CB|[_]Intergenic|[_]non-targeting|[_]TFEC|[_]MITF|[_]MTOR|[_]Akt1','', rownames(norm.mat)) %in% sigHits,
                   grep('CTRL1', colnames(norm.mat), invert=T)]

# log transform to go to logscale
submat <- log2(submat + 1)

# median sweep of vals
submat <- sweep(submat, 1, apply(submat, 1,median, na.rm=T))

# create annotation
ann <- data.table(group=str_extract(colnames(submat), 'BAFA1|Lib|DOXO|CTRL[123]|STARVE|TORIN|GPN|plasmid_BE|gRNA_noBE'),
                  cas9=ifelse(grepl('Abe8e|Bemax', colnames(submat)), str_extract(colnames(submat), 'Abe8e|Bemax'), 'Ctrl'))

colours <- list('cas9' = c('Abe8e' = '#4477AA', 'Bemax'="#228833", 'Ctrl'="#CCBB44"),
                'group' =c('BAFA1' = col.pal[1], 'Lib'=col.pal[2], 'DOXO'=col.pal[3], 'CTRL1'=col.pal[4], 'CTRL2'=col.pal[5], 'CTRL3'=col.pal[6],
                           'STARVE' = col.pal[7], 'TORIN'=col.pal[8], 'GPN'=col.pal[9], 'plasmid_BE'=col.pal[10], 'gRNA_noBE'=col.pal[11]))

colAnn <- HeatmapAnnotation(df = ann, col = colours)

# weirdly, our sample clustering isnt great
hm <- Heatmap(submat, name='log2(normCounts)/median', 
              cluster_rows = clusterWNA(submat, method='pearson'),
              na_col = 'grey', 
              #col=colorRamp2(breaks=c(-100,0,100), colors = c('blue', 'white', 'red')),
              show_row_names = F,  
              top_annotation = colAnn)
hm
BackupAsPDF(hm, 'tmoNormCounts.sigHits.heatmap',  dimensions=c(10,8))
```
Look at the logNorm counts
Ignore this....

```{r, eval=FALSE}

norm.mat <-  dcast(counts.dt, sgRNA_Gene~sample, value.var='CPM.normCounts') %>% 
  as.matrix(rownames='sgRNA_Gene')

submat <- norm.mat[gsub('[_]TFEB|[_]COLN1|[_]TFE3|[_]PPP3CB|[_]Intergenic|[_]non-targeting|[_]TFEC|[_]MITF|[_]MTOR|[_]Akt1','', rownames(norm.mat)) %in% sigHits,
                   grep('CTRL1', colnames(norm.mat), invert=T)]

# median sweep of vals
submat <- sweep(submat, 1, apply(submat, 1,median, na.rm=T))

# create annotation
ann <- data.table(group=str_extract(colnames(submat), 'BAFA1|Lib|DOXO|CTRL[123]|STARVE|TORIN|GPN|plasmid_BE|gRNA_noBE'),
                  cas9=ifelse(grepl('Abe8e|Bemax', colnames(submat)), str_extract(colnames(submat), 'Abe8e|Bemax'), 'Ctrl'))

colours <- list('cas9' = c('Abe8e' = '#4477AA', 'Bemax'="#228833", 'Ctrl'="#CCBB44"),
                'group' =c('BAFA1' = col.pal[1], 'Lib'=col.pal[2], 'DOXO'=col.pal[3], 'CTRL1'=col.pal[4], 'CTRL2'=col.pal[5], 'CTRL3'=col.pal[6],
                           'STARVE' = col.pal[7], 'TORIN'=col.pal[8], 'GPN'=col.pal[9], 'plasmid_BE'=col.pal[10], 'gRNA_noBE'=col.pal[11]))

colAnn <- HeatmapAnnotation(df = ann, col = colours)

# weirdly, our sample clustering isnt great
hm <- Heatmap(submat, name='log2(normCounts)/median', 
              cluster_rows = clusterWNA(submat, method='pearson'),
              na_col = 'grey', 
              col=colorRamp2(breaks=c(-20,0,20), colors = c('blue', 'white', 'red')),
              show_row_names = F,  
              top_annotation = colAnn)
hm
BackupAsPDF(hm, 'logNormCounts.sigHits.heatmap',  dimensions=c(10,8))
```

Scatterplots of log2FC correlations
Do all by all LFC correlation plots; seems some of the treatments may be more similiar than the cas9 editor

```{r}
library(ggpubr) # could also use geom_annotate..

# do a all by all scatterplot comparison
contrast.pairs <-  UniqueFactorCombos(sgRNA.dt$contrast, allow.dups = F, sep=',')
contrast.pairs

contrast.pairs[Var1 %like% 'GPN' & Var2 %like% 'BAFA1',]

apply(contrast.pairs,1,  function(x){
  
  print(sgRNA.dt[contrast %in% x, unique(contrast)])
  
  dt <- dcast(sgRNA.dt[contrast %in% x,], sgrna+Gene~contrast, value.var = c('LFC','sig'))
  # print
  setnames(dt, new=c('sgRNA', 'gene', 'log2FC.x', 'log2FC.y', 'sig.x', 'sig.y'))

  
  # could try nested if else but not working v well....
  dt[, label := 'not']
  dt[sig.x %in% c('up','down') & !sig.y %in% c('up','down'), label :=  'x-axis sig']
  dt[!sig.x %in% c('up','down') & sig.y %in% c('up','down'), label :=  'y-axis sig']
  dt[sig.x %in% c('up','down') & sig.y %in% c('up','down'), label :=  'both']
  
  g <- ggplot(dt, aes(x=log2FC.x, y=log2FC.y, color=label, label=sgRNA)) +
    geom_point() +
    geom_vline(xintercept = c(-1,1), linetype=2, alpha=0.4) +
    geom_hline(yintercept = c(-1,1), linetype=2, alpha=0.4) +
    scale_color_manual(values=c('y-axis sig'='#990033', 'x-axis sig'="#2A788EFF", 'not'='grey', 'both'="#440154FF")) +
    ggrepel::geom_text_repel(data=dt[label != 'not',],size = 2, max.overlaps = 20, show.legend = F) +
    xlab(paste(x[1], 'log2 fold change',sep=' ')) +
    ylab(paste(x[2], 'log2 fold change',sep=' ')) +
    ggtitle(paste('Fold Change Comparisons')) + 
    theme_bw()
  
  #add cor score
  g <- g + stat_cor(data=dt, aes(x=log2FC.x, y=log2FC.y, label=..r.label..), method='pearson', inherit.aes = F)
  
  BackupAsPDF(g, paste0('scatterplots/',x[1], 'vs', x[2],'.log2FC.scatterplots.'))
})
```

Lets look at overlap in sgRNA between the treatment groups and the editors
Looking at the heatmaps, I doubt we will get much overlap in the different editors

```{r}
# get abe8e and bemax sig groups seperately
sigSgs <- sgRNA.dt[sig != 'not', .(sgrna, Gene, contrast)] %>% 
  split(.$contrast)

sigSgs <- lapply(sigSgs, function(x){
  return(paste(x$sgrna, x$Gene, sep='-'))
})

abe8e.sig <- sigSgs[grep('Abe8e', names(sigSgs))]
bemax.sig <- sigSgs[grep('Abe8e', names(sigSgs), invert = T)]

```
Question why are we using a different control set for the DOXO comparison and is this relevant?

```{r}
names(bemax.sig) <- gsub('Bemax_|BemaxSerum_|Bemax', '', names(bemax.sig))
names(abe8e.sig) <- gsub('Abe8e_|Abe8eSerum_|Abe8e', '', names(abe8e.sig))

p <- plot(euler(bemax.sig),
          adjust_labels = TRUE,
          legend=F,
          labels=list(cex=.6, fontface=2),
          main = list(label='Bemax Cas9 Group Comparison Overlaps (Log2FC >1 & p.adj < 0.05)', cex=1, fontface=1),
          quantities = TRUE
     )
p
BackupAsPDF(p, 'vennDiagrams/bemax.allSigComparisons.vennDiagram', dimensions = c(8,8))

p <- plot(euler(abe8e.sig),
          adjust_labels = TRUE,
          legend=F,
          labels=list(cex=.6, fontface=2),
          main = list(label='Abe8e Cas9 Group Comparison Overlaps (Log2FC >1 & p.adj < 0.05)', cex=1, fontface=1),
          quantities = TRUE
     )
p
BackupAsPDF(p, 'vennDiagrams/abe8e.allSigComparisons.vennDiagram', dimensions = c(8,8))
```
Look at up-reg and down-reg separately

Get the sets
Seperate the names correctly
```{r}
# get abe8e and bemax sig groups seperately
sigSgs <- sgRNA.dt[sig != 'not', .(sgrna, Gene, contrast, sig)] %>% 
  split(.$contrast)



upSgs <- lapply(sigSgs, function(x){
    up.set <- x[sig == 'up', .(paste(sgrna, Gene, sep='-'))]$V1
    return(up.set)
})

downSgs <- lapply(sigSgs, function(x){
    down.set <- x[sig == 'down', .(paste(sgrna, Gene, sep='-'))]$V1
    return(down.set)
})

# upreg
abe8e.up <- upSgs[grep('Abe8e', names(upSgs))]
bemax.up <- upSgs[grep('Abe8e', names(upSgs), invert = T)]

#downreg
abe8e.down <- downSgs[grep('Abe8e', names(downSgs))]
bemax.down <- downSgs[grep('Abe8e', names(downSgs), invert = T)]


# fix names 
names(bemax.up) <- gsub('Bemax_|BemaxSerum_|Bemax', '', names(bemax.up))
names(abe8e.up) <- gsub('Abe8e_|Abe8eSerum_|Abe8e', '', names(abe8e.up))
names(bemax.down) <- gsub('Bemax_|BemaxSerum_|Bemax', '', names(bemax.down))
names(abe8e.down) <- gsub('Abe8e_|Abe8eSerum_|Abe8e', '', names(abe8e.down))

# add to list
sigSets <-  list('Abe8e UpReg' = abe8e.up,
                 'BeMax UpReg' = bemax.up,
                 'Abe8e DownReg' = abe8e.down, 
                 'BeMax DownReg' = bemax.down
)
```

Volcanoplot of the sig hits

Facet by genomic location
```{r}
lapply(unique(sgRNA.dt$contrast), function(x){
  
  subdt <- sgRNA.dt[contrast == x,]
  
  g <- ggplot(subdt, aes(x=LFC, y=-log10(FDR), color=sig, label=sgrna)) +
  geom_point() +
  geom_vline(xintercept = c(-1,1), linetype=2, alpha=0.4) +
  geom_hline(yintercept = -log10(0.05), linetype=2, alpha=0.4) +
  geom_text_repel(data=subdt[sig !='not'], show.legend = F, size=2, max.overlaps = 20) +
  facet_wrap(~Gene, ncol=3, scales='free_y') +
  ggtitle(x) +
  ylab('-log10 adjusted p-value') +
  xlab('Log Fold Change') +
  scale_color_manual(values=c('up'='#990033', 'down'='#336699', 'not'='grey')) +
  theme_classic() 
  
  BackupAsPDF(g, paste0('volcanoplots_individual/genomic_location/',x ,'.volcanoplots'), dimensions=c(12,14))
})


lapply(unique(sgRNA.dt$contrast), function(x){
  
  subdt <- sgRNA.dt[contrast == x,]
  subdt[, adj.pval := -log10(FDR)]
  subdt[ adj.pval > 20, adj.pval := 20 ]
  
  g <- ggplot(subdt, aes(x=LFC, y=adj.pval, color=sig, label=sgrna)) +
  geom_point() +
  geom_vline(xintercept = c(-1,1), linetype=2, alpha=0.4) +
  geom_hline(yintercept = -log10(0.05), linetype=2, alpha=0.4) +
  geom_text_repel(data=subdt[sig !='not'], show.legend = F, size=2, max.overlaps = 20) +
  facet_wrap(~Gene, ncol=3) +
  ggtitle(x) +
  ylab('-log10 adjusted p-value') +
  xlab('Log Fold Change') +
  scale_color_manual(values=c('up'='#990033', 'down'='#336699', 'not'='grey')) +
  theme_classic() 
  
  BackupAsPDF(g, paste0('volcanoplots_individual/genomic_location_fixedYaxis/',x ,'.volcanoplots'), dimensions=c(12,14))
})
```

Color by genomic location 
--
Not very informative; no real clustering by genomic breakdown, but maybe look at annotating heatmap
```{r}
col.pal <- randomcoloR::distinctColorPalette(k=10)

lapply(unique(sgRNA.dt$contrast), function(x){
  
  subdt <- sgRNA.dt[contrast == x,]
  subdt[, adj.pval := -log10(FDR)]
  subdt[ adj.pval > 20, adj.pval := 20 ]
  
  g <- ggplot(subdt, aes(x=LFC, y=(adj.pval), label=sgrna)) +
  geom_point(color='grey') +
  geom_point(data=subdt[sig != 'not' & contrast == x,], aes(x=LFC, y=adj.pval, color=Gene)) +
  geom_vline(xintercept = c(-1,1), linetype=2, alpha=0.4) +
  geom_hline(yintercept = -log10(0.05), linetype=2, alpha=0.4) +
  geom_text_repel(data=subdt[sig !='not'], show.legend = F, size=2, max.overlaps = 20) +
  ggtitle(x) +
  ylab('-log10 adjusted p-value') +
  xlab('Log Fold Change') +
  scale_color_manual(values=col.pal) +
  theme_classic() 
  
  g
  
  BackupAsPDF(g, paste0('volcanoplots_individual/col_by_genomic_location/',x ,'.volcanoplots'))
})
```

```{r}
sgRNA.dt[, LFC.zscore := scale(LFC, center=T, scale=T), by=.(contrast)]
sgRNA.dt %>%  summary()

g <- ggplot(sgRNA.dt, aes(x=LFC, y=Gene, fill=Gene)) +
  geom_density_ridges() +
  ggtitle('Log2FC distributions') +
  geom_vline(xintercept = c(-1,1), linetype=2, alpha=0.4) +
  scale_fill_manual(values=col.pal) +
  facet_wrap(~contrast, scales='free_x',  ncol=3) +
  theme_bw() 
g
#BackupAsPDF(g, 'ridgeplots/LFCdistributions.ridgeplot', dimensions = c(12,16))
```
Interesting that basically all genes are being upregulated in response to treatment
Check the raw and normalized counts; look fin

```{r}
counts.dt[sgRNA == 'BE_05' & group %in% c('Abe8e_BAFA1', 'Abe8eDMSO_CTRL2', 'Abe8eDMSO_CTRL1')]
```
Variant Annotation
------
Annotate the sgRNA by mutational consequence
Read in the VEP annotated data

```{r}
abe8e.canonical <- fread('./160424_annotateVariants_data/2024_04_17_abe8ecanonicalVEPAnnotations.txt')
bemax.canonical <- fread('./160424_annotateVariants_data/2024_04_17_bemaxcanonicalVEPAnnotations.txt')

# severe set 
abe8e.severe <- fread('./160424_annotateVariants_data/2024_04_17_abe8ehighestImpactVEPAnnotations.txt')
bemax.severe <- fread('./160424_annotateVariants_data/2024_04_17_bemaxhighestImpactVEPAnnotations.txt')


vep.anno.list <- list('abe8e.canonical' = abe8e.canonical,
                      'bemax.canonical' = bemax.canonical,
                      'abe8e.severe' = abe8e.severe,
                      'bemax.severe' = bemax.severe)

lapply(vep.anno.list, function(x){
  x[, c('sgRNA_sequence', 'windowEdits') := tstrsplit(Name,'__', keep=c(1,2))]
})


# seems we have lost many sgRNA..
# need to go back to the VEP data and double check this... why did we lose so many sgRNA?
lapply(vep.anno.list, function(x){
  x[,unique(sgRNA_sequence)] %>%  length()
})
```
```{r}
# merge vep anno with cripsr lib info to get sgRNA name
# just map by sequence... risky? Should be fine these are 20bp...
crispr.lib <- fread('./docs/crispr.lib.txt', header = F)
setnames(crispr.lib, new=c('sgRNA', 'sgRNA_sequence', 'Gene'))

g <- ggplot(crispr.lib[,.N, by=.(Gene)], aes(x=Gene, y=N, fill=Gene)) +
  geom_bar(stat='Identity') +
  ggtitle('sgRNAs per gene') +
  scale_fill_manual(values=col.pal) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90))

BackupAsPDF(g, 'NsgRNA.perGene.barplot', dimensions=c(6,6))
  
unique(crispr.lib, by='sgRNA_sequence')
crispr.lib$sgRNA_sequence %>%  unique() %>%  length()

dup.ids <- crispr.lib[duplicated(sgRNA_sequence), sgRNA_sequence]

#fwrite(crispr.lib[sgRNA_sequence %in% dup.ids,][order(sgRNA_sequence)], ScriptAndDatedFileName('duplicate.sgRNA.txt'))
crispr.lib[sgRNA_sequence %in% dup.ids,][order(sgRNA_sequence)]
```

Look at the CRISPR library information
Do these sgRNA have different target sites?

```{r}
guideContext.dt <- list(TFEB = read_xlsx(path='./docs/base editing gRNA location_4.9.24.xlsx', sheet=2),
  TFE3 = read_xlsx(path='./docs/base editing gRNA location_4.9.24.xlsx', sheet=3),
  TFEC = read_xlsx(path='./docs/base editing gRNA location_4.9.24.xlsx', sheet=4),
  MITF = read_xlsx(path='./docs/base editing gRNA location_4.9.24.xlsx', sheet=5),
  MTOR = read_xlsx(path='./docs/base editing gRNA location_4.9.24.xlsx', sheet=6),
  AKT1 = read_xlsx(path='./docs/base editing gRNA location_4.9.24.xlsx', sheet=7),
  COLN1 = read_xlsx(path='./docs/base editing gRNA location_4.9.24.xlsx', sheet=8),
  PP3CB = read_xlsx(path='./docs/base editing gRNA location_4.9.24.xlsx', sheet=9)) %>%  
  rbindlist(idcol=T, fill=T)


# so some of these records have been entered twice.. not a big problem, lets just ensure we are dropping these on the merge
guideContext.dt[Guide %in% dup.ids,][order(Guide)]
```
# merge the files
Approx same number of extract records in each... is a guide duplicated?
```{r}
vep.cols.oi <- c("Name","Feature","Consequence","Location","REF_ALLELE","Allele","IMPACT","SYMBOL","Gene",
                 "cDNA_position","CDS_position","Protein_position","Amino_acids","Codons",'sgRNA_sequence', 'windowEdits',
                 "EXON","INTRON","HGVSc","HGVSp","SWISSPROT","MANE_SELECT", "Existing_variation")
                 
                 
# subset to the cols we want
vep.anno.list <- lapply(vep.anno.list, function(dt){
  dt[, ..vep.cols.oi]
})


# merge the annotation list with the crispr IDs 
vep.anno.list <- lapply(vep.anno.list, function(dt){
  merge(x=dt, y=crispr.lib[,-c('Gene')], by='sgRNA_sequence', all.x=T)
})
```
Find and remove duplicate rows
Leave in for now... just be aware of this and provide the IDs for Ronald
```{r}
lapply(vep.anno.list, function(dt){
  dup.rows <- dt[duplicated(sgRNA_sequence), sgRNA_sequence]
  dt[sgRNA_sequence %in% dup.rows][order(sgRNA_sequence)]
})
```
Out of curisoity, how many of our sgRNA were annotated and what was their annotation?
Do these annotations make sense? If many of these are missense, would we also not expect to see a significant amount of mutations resulting in LOF?

```{r}
canonical.vep <- vep.anno.list[grepl('canonical', names(vep.anno.list))] %>% 
  rbindlist(idcol=T)

col.pal <- randomcoloR::distinctColorPalette(k=length(unique(canonical.vep$Consequence)))

canonical.vep[, IMPACT := factor(IMPACT, levels=c('MODIFIER','LOW','MODERATE','HIGH'))]

g <- ggplot(canonical.vep[,.N, by=.(Consequence, .id)], aes(x=Consequence, fill=Consequence, y=N)) +
  geom_bar(stat='Identity') +
  ggtitle('Counts of SO Mutational Edit Consequences') +
  facet_wrap(~.id) +
  theme_classic() +
  scale_fill_manual(values=col.pal) +
  theme(axis.text.x = element_text(angle=90, size=7))

BackupAsPDF(g, paste0('variant_breakdown/canonicalVariantsConsequenceCounts'), dimensions=c(8,7))

# breakdown by mutant severity..
# If many of these are deleterious, does our normalisation proceedure hold?

g <- ggplot(canonical.vep[,.N, by=.(IMPACT, Consequence, .id)], aes(x=IMPACT, fill=Consequence, y=N)) +
  geom_bar(stat='Identity') +
  ggtitle('Mutational Edit Consequences (SO terms)') +
  xlab('Sequence Ontology Mutational Consequence') +
  ylab('N sgRNA') +
  facet_wrap(~.id) +
  theme_classic() +
  scale_fill_manual(values=col.pal) +
  theme(axis.text.x = element_text(angle=90, size=7))

g
BackupAsPDF(g, paste0('variant_breakdown/canonicalVariantsImpact.colSOterm.barplot'), dimensions=c(8,7))
```
Lets look in the different groups; do we see more of the severe mutations in the genes with more DEGs?
Looks like most of these modifiers are in MTOR.. this site also seems to be the most 

```{r}

g <- ggplot(canonical.vep[,.N, by=.(IMPACT, Consequence, SYMBOL, .id)], aes(x=IMPACT, fill=Consequence, y=N)) +
  geom_bar(stat='Identity') +
  ggtitle('Mutational Edit Consequences (SO terms)') +
  xlab('Sequence Ontology Mutational Consequence') +
  ylab('N sgRNA') +
  facet_grid(.id~SYMBOL) +
  theme_classic() +
  scale_fill_manual(values=col.pal) +
  theme(axis.text.x = element_text(angle=90, size=7),
        strip.text.x = element_text(size=7))

g
BackupAsPDF(g, 'mutationalConsequence.geneBreakdown.barplot', dimensions=c(10,6))
```
Merge the VEP annotation with output from MAGECK DE testing

```{r}
# before merging, we need to adjust the gene names for MCOLN1 and AKT1
sgRNA.dt[Gene == 'Akt1', Gene := 'AKT1']
sgRNA.dt[Gene == 'COLN1', Gene := 'MCOLN1']

# combined this way as 1:1 mapping when joing res and cas9 variant tables seperately
merged.LFC.list <- lapply(vep.anno.list, function(dt){
 LFC.dt <- merge(x=sgRNA.dt, y=dt, by.x=c("sgrna","Gene"), by.y=c('sgRNA', "SYMBOL"), all.x=T)
 return(LFC.dt)
})
 
canonical.dt <- merged.LFC.list[grepl('canonical', names(merged.LFC.list))] %>% 
  rbindlist(idcol=T)
```
What is the NA VEP annotation?
These are the missing sg... drop these from the plots...

```{r}
# look up this list of sgRNA in our vep results and lets see why they were dropped..
# 
fwrite(canonical.dt[is.na(IMPACT),], ScriptAndDatedFileName('missingSgRNAs.txt'))
fwrite(canonical.dt[!Gene %in% c('Intergenic', 'non-targeting') & is.na(IMPACT),], ScriptAndDatedFileName('missingSgRNAs.noIntragenicOrNonTargeting.txt'))

```

lets look at the z-score distributions by mutant severity
```{r}
# per contrast get the z-scoress of the log2FC distributions
canonical.dt[, LFC.zscore := scale(LFC, center=T, scale=T), by=.(contrast, .id)]

canonical.dt[, comparison := gsub('Abe8e[_]{0,1}|Bemax[_]{0,1}','', contrast)]
contrasts.oi <- unique(canonical.dt$comparison)

# for the intergenic/non-targeting, add these as an IMPACT type to view their distributions
canonical.dt[Gene %in% c('Intergenic', 'non-targeting'), IMPACT := Gene]

genes.oi <- unique(canonical.dt$Gene)
genes.oi <- genes.oi[-c(5,6)] # drop the controls
```

```{r}
lapply(contrasts.oi, function(x){
  
  lapply(genes.oi, function(y){
    
    # need to make a new 'plot' dt with the intergenic group
    dt <- canonical.dt[!is.na(IMPACT) & comparison == x & Gene == y,]
    ctrl.dt <- canonical.dt[Gene %in% c('Intergenic', 'non-targeting'),]
    new.dt <- rbind(dt, ctrl.dt)
    new.dt[, IMPACT := factor(IMPACT, levels=c('non-targeting', 'Intergenic', 'MODIFIER', 'LOW','MODERATE','HIGH'))]
    new.dt[, .id := gsub('[.]canonical', '', .id)]
  
    g <- ggplot(new.dt, aes(x=LFC.zscore, y=IMPACT, fill=IMPACT, height = after_stat(density))) +
      geom_density_ridges(scale = 1.3, stat='density') +
      xlab('z-score (Log2FC)') +
      ylab('Mutation Consequence') +
      ggtitle(paste0(x, ' ', y)) +
     #scale_fill_manual(values=col.pal) +
      scale_fill_brewer(palette = 4) +
      facet_grid(~.id) +
      theme_ridges() 

  BackupAsPDF(g, paste0('ridge/', y, '_', x,'.LFC.zscoreDistributions.mutantConsequence.ridgeplot'), dimensions=c(10,8))
  })
})
```



Boxplots of log2FC values by impact/gene
Do the mutants with more severity have a wider distribution?

```{r}
lapply(contrasts.oi, function(x){
  
    # need to make a new 'plot' dt with the intergenic group
    dt <- canonical.dt[!is.na(IMPACT) & comparison == x & Gene %in% genes.oi,]
    dt[, IMPACT := factor(IMPACT, levels=c('non-targeting', 'Intergenic', 'MODIFIER', 'LOW','MODERATE','HIGH'))]
    dt[, .id := gsub('[.]canonical', '', .id)]
  
    g <- ggplot(dt, aes(x=IMPACT, y=LFC, fill=IMPACT)) +
      geom_boxplot(outliers=F) +
      xlab('Mutation Consequence') +
      ylab('Log2 FC') +
      ggtitle(paste0(x)) +
      scale_fill_brewer(palette = 4, type='qual') +
      facet_grid(.id~Gene) +
      theme_classic() +
      theme(axis.text.x = element_text(angle=90))
  g
  
  BackupAsPDF(g, paste0('boxplots/', x,'.LFC.Distributions.mutantConsequence.faceted.boxplot'), dimensions=c(9,14))
  
})
```

Plot, per contrast the distribution of LFC along the protein position & facet by gene and cas9
Idea is here to highlight regions that may have a lot of mutations

```{r}
canonical.dt[, startSite := (tstrsplit(Location, ':|-', keep=2))]
canonical.dt[, startSite := as.numeric(startSite)]

115984475 > 7522731

lapply(contrasts.oi, function(x){
  
  lapply(genes.oi, function(y){
    
    # need to make a new 'plot' dt with the intergenic group
    dt <- canonical.dt[!is.na(IMPACT) & comparison == x & Gene %in% y,]
    dt[, IMPACT := factor(IMPACT, levels=c('MODIFIER', 'LOW','MODERATE','HIGH'))]
    dt[, cas9 := gsub('[.]canonical', '', .id)]
    
    g <- ggplot(dt, aes(x=reorder(startSite, startSite, decreasing=F), y=LFC, shape=cas9, fill=IMPACT, label=sgrna)) +
      geom_point() +
      geom_text_repel(data=dt[sig != 'not',], size=2, max.overlaps = 20) +
      xlab('Edit Window Start Site') +
      ylab('Log2 FC') +
      ggtitle(paste0(y, ' Contrast', x)) +
      scale_shape_manual(values = c(21,24)) +
      scale_fill_manual(values=c('MODIFIER'='grey', 'LOW'='lightyellow', 'MODERATE'='orange', 'HIGH'='darkred')) +
      theme_classic() +
      theme(axis.text.x=element_blank(),
            axis.ticks.x = element_blank()) +
      guides(fill = guide_legend(override.aes = list(shape =21) ) ,
         color = guide_legend(override.aes = list(shape =21) ) )
 
  g
  BackupAsPDF(g, paste0('dotplots_labelled/',y, '_', x,'.LFC.mutantConsequence.windowStartSite.dotplot'), dimensions=c(10,6))
  })
})

```

Write out the canonical variant annotation table and share with 



###
Not used...
###

# now merge this annotation data with the LFC results
# we want to keep all of our log2FC results
These results will only be for our genes, intergenic and 

```{r}
merged.LFC.list <- lapply(vep.anno.list, function(dt){
 LFC.dt <- merge(x=sgRNA.dt, y=dt, by.x="sgrna", by.y='sgRNA', all.x=T)
 return(LFC.dt)
})
```

Now we want to 'split' this list to canonical and mutant groups
Then rename the 'consequence' col; if empty, name it as the gene
we can then filter these by gene name to see what we miss; majority should be intergenic/off-target hopefully....

```{r}
canonical.dt <- merged.LFC.list[grepl('canonical', names(merged.LFC.list))] %>% 
  rbindlist(idcol=T)



canonical.dt[,.N, by=.(Consequence,.id)]

ggplot(canonical.dt, aes(x=Consequence,)
```

Todo
---
Ridge Plots with variant annotation 
Plots of log2FC with variant on x by location
Heatmap of log2FCs with genomic location and severity of the variants annotated? Also annotate how many comparisons it was sig in (seperate barplots for both cas9)

Ask Ronald about which contrasts he is interested in and we can fish out these sgRNA;
Also you could plot all sgRNA sig in more than one contrast (matrix maybe?)

plot the sets

```{r}
lapply(seq_along(sigSets), function(x,n,i){
  
  p <- plot(euler(x[[i]]),
          adjust_labels = TRUE,
          legend=F,
          labels=list(cex=.6, fontface=2),
          main = list(label=paste0(n[[i]], ' Comparison Overlaps (Log2FC >1 & p.adj < 0.05)'), cex=1, fontface=1),
          quantities = TRUE
     )
  p
  BackupAsPDF(p, paste0('vennDiagrams/', n[[i]], '.Comparisons.vennDiagram'), dimensions = c(8,8))
}, x=sigSets, n=names(sigSets))
```

Overlap in cas9 editors