---
title: "VariantAnnotation"
author: "Martin Gordon"
date: "2023-11-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## VariantAnnotation pkg tutorial

workflow for annotating and filtering genetic varianting using `VariantAnnotation`

packages
```{r}
library(ggplot2)
library(data.table)
library(VariantAnnotation)
```

test data from `vcfR` tutorial

```{r}
pkg <- "pinfsc50"
vcf_file <- system.file("extdata", "pinf_sc50.vcf.gz", package = pkg)
dna_file <- system.file("extdata", "pinf_sc50.fasta", package = pkg)
gff_file <- system.file("extdata", "pinf_sc50.gff", package = pkg)

vcf_file
```


```{r}
vcf_file <- system.file("extdata", "chr22.vcf.gz", package="VariantAnnotation")
vcf <- readVcf(vcf_file, 'hg19')

readVcf()
```
extract header information with `header` ( extract subsections)
```{r}
header(vcf)
```
parsing genomic positions
`rowRanges` contains information from the CHROM, POS, and ID fields of the VCF file,

```{r}
rowRanges(vcf)

# can access specific elements using 
ref(vcf)
alt(vcf)

# or look at S4 slots and extract...
rowRanges(vcf) %>%  str()
rowRanges(vcf)$REF

```
parsing genotype data

```{r}
geno(vcf)
geno(header(vcf))['DS',]

# variant location by sample matrix; each row is variant location and each cell with have a value for it 
geno(vcf)$DS
```

locate variants around genes using `locateVariants`

```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

seqlevels(vcf) <- "chr22" # For overlap methods to work properly the chromosome names (seqlevels) must be compatible in the objects being compared
rd <- rowRanges(vcf)
loc <- locateVariants(rd, txdb, CodingVariants())
head(loc, 3)
```
queries: more variants in any genes?

```{r}
## Did any coding variants match more than one gene?
# 15 gene have more than one gene
splt <- split(mcols(loc)$GENEID, mcols(loc)$QUERYID) 
table(sapply(splt, function(x) length(unique(x)) > 1))


## Summarize the number of coding variants by gene ID.
splt <- split(mcols(loc)$QUERYID, mcols(loc)$GENEID)
sapply(splt, function(x) length(unique(x)))
```

amino acid changes
`predictCoding` computes amino acid coding changes for non-synonymous variants

Only ranges in query that overlap with a coding region in the subject are considered. 
Reference sequences are retrieved from either a BSgenome or fasta file specified in seqSource
```{r}
library(BSgenome.Hsapiens.UCSC.hg19) # need to load but can use fasta for seq source
coding <- predictCoding(vcf, txdb, seqSource=)

```
Can also use SIFT and PolyPhen functionality to assess disruption of variant


