---
title: "vcfProcessing"
author: "Martin Gordon"
date: "2023-11-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Create and test some R functions for processing VCF files 

vcfR tool seems to be useful for filtering variants based on quality scores etc and writing output to new file

Need:
- Identify the amino acid substitutions from the vcf
- The number of reads associated with each UMI-barcode
- barcode groups for cells with the same AA substitution



To verify variant calls for a specific codon, we compared each barcode group among all sample VCFs. The minor fraction of variant calls in a particular barcode group that did not agree with a designed codon substitution was filtered and discarded.

Frequencies of reads were calculated for each barcode group in each sample. Read counts supporting the codon substitution were tallied from the flow sorted sequencing files. Read counts were normalized to RPM. Transcriptional activities were calculated by comparing the distributions of read counts across flow sorted samples across all barcodes for a particular variant (see data file 1). Activities were calculated for each mutant using formula 1 below.


Packages
```{r}
library(vcfR)
library(data.table)
library(ggplot2)
```


1. vcfR package

test data..
```{r}
pkg <- "pinfsc50"
vcf_file <- system.file("extdata", "pinf_sc50.vcf.gz", package = pkg)
dna_file <- system.file("extdata", "pinf_sc50.fasta", package = pkg)
gff_file <- system.file("extdata", "pinf_sc50.gff", package = pkg)
```

read in files
```{r}
#S4 object with 3 slots
vcf <- vcfR::read.vcfR(vcf_file, verbose = T)

dna <- ape::read.dna(dna_file, format='fasta')
gff <- fread(gff_file)
```

This function creates achromR object
plot gives an overview of the key statistics
```{r}
chrom <- create.chromR(name='Supercontig', vcf=vcf, seq=dna, ann=gff)

plot(chrom) #no SNP densities plot

# after viewing the data can filter
# filtering actually just createzs a logical vector, so this is reversible
chrom <- masker(chrom, min_QUAL = 1, min_DP = 300, max_DP=700, min_MQ=59.8, max_MQ=60.1)
plot(chrom) # all this info lies in var.info obj
```
process the vcf for further inspection
This function calls furtehr helper functions to process variant sequence and annotation data for visualisation

```{r}
chrom <- proc.chromR(chrom, verbose=TRUE, )
chromoqc(chrom)
```
if prefer tidyberse, convert to tibble (list of tibbles)

```{r}
tidy.vcf <- vcfR2tidy(vcf)
```
extracting matrices from vcfR objects

```{r}
head(vcf)
gt <- extract.gt(vcf) #format info is stored in the meta section
gt <- extract.gt(vcf, element = 'DP', as.numeric = TRUE) # can also extract specific section of the vcf
```


after filtering, you can write out the processed vcf file
```{r}
#write.vcf()
```



