---
title: "AF3_fullSacsin"
author: "Martin Gordon"
date: "2024-05-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Plots of the full sacsin template produced by AF3

```{r}
library(data.table)
library(magrittr)
library(ggplot2)
library(jsonlite) # convert json to do
library(stringr)
library(gggenes) # draw gene segements
library(patchwork) # combine plots
library(ComplexHeatmap)
library(circlize)
#library(rjson) # for reading in the R json files
library(jsonlite)
library(RColorBrewer)

source ("../../utils/bp_utils/UniprotIDMapping.R") #uniprot to gene id mapping
source("../../utils/bp_utils/manageScriptData.R")
```


read in the data for each of the models
```{r}
models <- dir(path='/Users/martingordon/Documents/projects/032524_AForget_ARSACS_AF/output/fold_sacsin_150524/', pattern="fold_sacsin_150524_full_data_[01234].json", full.names = T)

af.mods <- lapply(models, function(x){jsonlite::fromJSON(x) })
names(af.mods) <- str_extract(models, 'full_data_[01234]')
```

heatmap of pae scores

```{r}
col.pal <- brewer.pal(8, name='Dark2')
col.pal2 <- brewer.pal(8, name='Pastel2')

col.pal <- randomcoloR::distinctColorPalette(k=10)
lapply(af.mods, str)

# first annotation 
sacsin.anno <-rep('',4579)
sacsin.anno[1:67] <- 'UBL'
sacsin.anno[107:505] <- 'SRR1'
sacsin.anno[1471:1921] <- 'SRR2'
sacsin.anno[2539:2922] <- 'SRR3'
sacsin.anno[3659:3735] <- 'XPCB'
sacsin.anno[4295:4384] <- 'DNAJ'
sacsin.anno[4450:4466] <- 'HEPN'

# cloning anno
clone.anno <-rep('',4579)
clone.anno[1:84] <- 'ubl'
clone.anno[84:374] <- 'SIRPT1'
clone.anno[1444:2443] <- 'SIRPT2'
clone.anno[2512:4282] <- 'SIRPT3'
clone.anno[4316:4420] <- 'SacsJ'
clone.anno[4422:4579] <- 'HEPN'


sacsin.anno

lapply(seq_along(af.mods), function(x,n,i){

  # set up matrix
  mat <- x[[i]]$pae
  rownames(mat) <- x[[i]]$token_res_ids
  colnames(mat) <- x[[i]]$token_res_ids

  #simplify dim names; if not in 
  rownames(mat) <- ifelse(rownames(mat) %in% seq(1, nrow(mat), by=100), rownames(mat), '')
  colnames(mat) <- ifelse(rownames(mat) %in% seq(1, nrow(mat), by=100), rownames(mat), '')
  
  hm1 <- Heatmap(mat, 
                cluster_rows = FALSE, 
                cluster_columns = FALSE, 
                column_title = paste0('model ', gsub('full_data_', '', n[[i]])),
                show_row_names = T, 
                show_column_names = T,
                name='Predicted Aligned Error (Angstroms)')
 
   hm2 <- Heatmap(sacsin.anno, 
          name='Domains',
          col=structure(col.pal, names=unique(sacsin.anno)),
          width=unit(0.5, 'cm'))
  
   hm3 <- Heatmap(clone.anno, 
          name='Gentil 2018',
          col=structure(col.pal, names=unique(clone.anno)),
          width=unit(0.5, 'cm')) 
   
  BackupAsPDF(draw(hm3+hm2+hm1), paste0(n[[i]], '.pae.heatmap'), dimensions=c(20,20))
  
},x=af.mods, n=names(af.mods))
```

Lets cut this into smaller submats 

```{r}
#broadly define boundaries 
bond.start <- c(0, 1401, 2512, 3401)
bond.end <- c(1401,2511, 3400, 4579)


lapply(seq_along(af.mods), function(x,n,i){

  # set up matrix
  mat <- x[[i]]$pae
  rownames(mat) <- x[[i]]$token_res_ids
  colnames(mat) <- x[[i]]$token_res_ids

  #simplify dim names; if not in 
  rownames(mat) <- ifelse(rownames(mat) %in% seq(1, nrow(mat), by=20), rownames(mat), '')
  colnames(mat) <- ifelse(rownames(mat) %in% seq(1, nrow(mat), by=20), rownames(mat), '')
  
  # get submat
  lapply(seq_along(bond.start), function(z){
    
    
    # set up file to save 
    png(paste0('/Users/martingordon/Documents/projects/032524_AForget_ARSACS_AF/AF3_SacsinComplete_plots_data/pdfs/',n[[i]],'_', bond.start[z],'_', bond.end[z],'.pae.heatmap.png'), 
        width = 15, height = 15, 
        units = "in", res = 100)
    
    
    subMat <- mat[bond.start[z]:bond.end[z],
                 bond.start[z]:bond.end[z]]

  
    hm1 <- Heatmap(subMat, 
                cluster_rows = FALSE, 
                cluster_columns = FALSE, 
                column_title = paste0('model ', gsub('full_data_', '', n[[i]])),
                show_row_names = T, 
                show_column_names = T,
                name='Predicted Aligned Error (Angstroms)')
 
    hm2 <- Heatmap(sacsin.anno[bond.start[z]:bond.end[z]], 
          name='Domains',
          col=structure(col.pal, names=unique(sacsin.anno)),
          width=unit(0.5, 'cm'))
  
    hm3 <- Heatmap(clone.anno[bond.start[z]:bond.end[z]], 
          name='Gentil 2018',
          col=structure(col.pal2, names=unique(clone.anno)),
          width=unit(0.5, 'cm')) 
   
    #BackupAsPDF(draw(hm3+hm2+hm1), paste0(n[[z]],'_', bond.start[z],'_', bond.end[z],'.pae.heatmap'), dimensions=c(16,16))   
    draw(hm3+hm2+hm1)
    dev.off()
  })
  
},x=af.mods, n=names(af.mods))
  
```
},x=af.mods, n=names(af.mods))




testing 

```{r}
lapply(seq_along(af.mods), function(x,n,i){

  # set up matrix
  mat <- x[[i]]$pae
  rownames(mat) <- x[[i]]$token_res_ids
  colnames(mat) <- x[[i]]$token_res_ids

  #simplify dim names; if not in 
  rownames(mat) <- ifelse(rownames(mat) %in% seq(1, nrow(mat), by=100), rownames(mat), '')
  colnames(mat) <- ifelse(rownames(mat) %in% seq(1, nrow(mat), by=100), rownames(mat), '')
  

  hm1 <- Heatmap(mat, 
                cluster_rows = FALSE, 
                cluster_columns = FALSE, 
                column_title = paste0('model ', gsub('full_data_', '', n[[i]])),
                show_row_names = T, 
                show_column_names = T,
                name='Predicted Aligned Error (Angstroms)')
 
   hm2 <- Heatmap(sacsin.anno, 
          name='Domains',
          col=structure(col.pal, names=unique(sacsin.anno)),
          width=unit(0.5, 'cm'))
   hm1 + hm2
  #BackupAsPDF(draw(hm2 + hm1), paste0(n[[i]], '.pae.heatmap'), dimensions=c(14,14))
  
},x=af.mods, n=names(af.mods))
```



Read the summary results and plot the avg pLDDT per model

```{r}
summary <- dir(path='/Users/martingordon/Documents/projects/032524_AForget_ARSACS_AF/output/fold_sacsin_150524/', pattern="fold_sacsin_150524_summary_confidences_[01234].json", full.names = T)

summary.dt <- lapply(summary, function(x){jsonlite::fromJSON(x) })
names(summary.dt) <- str_extract(summary, 'summary_confidences_[01234]')
summary.dt <- rbindlist(summary.dt, idcol = 'model')

```

plot of avg ptm

```{r}
summary.dt[,model := gsub('summary_confidences_', '', model)]

g <- ggplot(summary.dt, aes(x=model, y=ptm, fill=model)) +
  geom_bar(stat = 'Identity') +
  ggtitle('SACSIN model avg pTM scores') +
  geom_hline(yintercept = 0.5, linetype=2) +
  ylim(0,1) +
  scale_fill_brewer(type='qual',palette=2) +
  theme_bw()
g <- g + annotate(geom='text', y = 0.55, x=1.5, label='pTM confidence threshold')

BackupAsPDF(g, 'avg.ptm.barplots')
```


Do a linechart of the pLDDT plots

First need the scores per residue; read in the model.cif files


```{r}
model.paths <- dir('./output', pattern='*.cif', recursive = T, full.names = T)

# this is dependent on the size of the strucutre
model.list <- lapply(model.paths, function(x){
  fread(cmd = paste0("tail -n +9353 ", x))
})

names(model.list) <- str_extract(model.paths, 'model_[0-5]{1}')

model.dt <- rbindlist(model.list, idcol='model')
setnames(model.dt, new=c('model', 'group_PDB', 'id', 'type_symbol', 'label_atom_id','label_alt_id','label_comp_id', 'label_asym_id',
                         'label_entity_id','label_seq_id', 'pdbx_PDB_ins_code', 'x_coord', 'y_coord', 'z_coord', 'occupancy',
                         'B_iso_or_equiv','auth_seq_id', 'auth_asym_id','pdbx_PDB_model_num'))


# create an average pLDDT score per residue
model.dt[, pLDDT := mean(B_iso_or_equiv), by=.(model, auth_seq_id)]

# average score across the 5 models 
model.dt[, avg_pLDDT := mean(pLDDT), by=.(auth_seq_id)]
```

Create Sacsin DT for annotation

```{r}
sacsin.dt <- data.table(molecule=c('SACS','SACS','SACS','SACS','SACS','SACS','SACS','SACS','SACS','SACS'),
                        region=c('Domain1','Domain2','Domain3', 'UBL', 'SRR1', 'SRR2', 'SRR3','XPCB', 'DNAJ', 'HEPN'),
                        start=c(1,1,1,1,1,1,1,1,1,1),
                        end=c(4579,4579,4579,4579,4579,4579,4579,4579,4579,4579),
                        from=c(1, 1471, 2922,1,107,1471,2539,3659,4295,4450),
                        to=c(1471,2922,4466,67,505,1921,2922,3735,4384,4466))


sacsin.dt <- data.table(molecule=c('SACS','SACS','SACS','SACS','SACS','SACS','SACS','SACS','SACS','SACS',
                                   'SACS','SACS','SACS','SACS','SACS'),
                        region=c('Domain1','Domain2','Domain3', 'UBL', 'SRR1', 'SRR2', 'SRR3','XPCB', 'DNAJ', 'HEPN',
                                 'SIRPT1','SIRPT2','SIRPT3','Ubl','Hepn'),
                        start=c(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1),
                        end=c(4579,4579,4579,4579,4579,4579,4579,4579,4579,4579,4579,4579,4579,4579,4579),
                        from=c(1, 1471, 2922,1,107,1471,2539,3659,4295,4450,84,1444,2512,1,4422),
                        to=c(1471,2922,4466,67,505,1921,2922,3735,4384,4466,1374,2443,4282,84,4579))
```



Lets plot the avg pLDDT for all model

```{r}

g <- ggplot(model.dt, aes(x=auth_seq_id, y=avg_pLDDT, color=avg_pLDDT)) +
    geom_line(alpha=0.6, size=1) +
    geom_hline(yintercept = c(70,90), linetype=2, alpha=0.6) +
    annotate(geom='text',x=300, y=32, label=paste0('Best model mean pLDDT: ', model.dt[model == 'model_0',round(mean(pLDDT),1)]), size=5) +
    #annotate(geom='text',x=min.pos+40, y=92, label='excellent', color='black') +
    scale_color_gradientn("plDDT", limits = c(0,100), colors = c(red = "#FE0012",
                                                      orange = "#FFA42B",
                                                      yellow = "#FFFD42","#FFFD42",
                                                      palegreen = "palegreen2",
                                                      blue = "lightblue","lightblue",#"#6097E8",
                                                      darkBlue = "#001DF9"),
                        values = c(0, .5, 0.7,0.75,0.8,0.85, 0.9, 1.0)) +
    scale_y_continuous(breaks=seq(0,140, by=10)) +
    scale_x_continuous(breaks=seq(0, 4579, by=200)) +
    ggtitle(paste0('Sacsin AF3 avg pLDDT')) +
    ylab('mean pLDDT (5 models)') +
    xlab('Amino acid residue') +
    theme_bw() +
    guides(alpha=guide_legend(title="mean pLDDT"))
 g 
  # I need to modify the anno dt to only extend as far as the region listed
  
q <- ggplot(sacsin.dt[region %in% c('UBL', 'SRR1', 'SRR2', 'SRR3','XPCB', 'DNAJ', 'HEPN'),], aes(xmin=start, xmax=end, y=molecule)) +
    geom_gene_arrow(fill = "white") +
    geom_subgene_arrow(data = sacsin.dt[region %in% c('UBL', 'SRR1', 'SRR2', 'SRR3','XPCB', 'DNAJ', 'HEPN'),], 
                       aes(xmin = start, xmax = end, y = molecule, fill = region, xsubmin = from, xsubmax = to), color="black", alpha=.7) +
    scale_fill_brewer(type='qual', palette=2) +
    theme_genes()


p <- ggplot(sacsin.dt[region %in% c('SIRPT1','SIRPT2','SIRPT3','Ubl','Hepn'),], aes(xmin=start, xmax=end, y=molecule)) +
    geom_gene_arrow(fill = "white") +
    geom_subgene_arrow(data = sacsin.dt[region %in% c('SIRPT1','SIRPT2','SIRPT3','Ubl','Hepn'),], 
                       aes(xmin = start, xmax = end, y = molecule, fill = region, xsubmin = from, xsubmax = to), color="black", alpha=.7) +
    scale_fill_brewer(type='qual', palette=1) +
    theme_genes()
  
patch <- g/q/p + plot_layout(heights=c(6,1,1))
patch
BackupAsPDF(patch, 'sacsin.AF3.mean.plddt.linechart.Allgenetrack', dimensions = c(16,13))

```
```{r}
model.dt[model == 'model_0', .(pLDDT)]
model.dt[model == 'model_2', .(pLDDT)]
model.dt[model == 'model_0', .(avg_pLDDT)]
```
plDDT all models
```{r}
lapply(af.mods, str)

g <- ggplot(model.dt, aes(x=auth_seq_id, y=pLDDT, color=model)) +
    geom_line(alpha=0.6, size=1) +
    geom_hline(yintercept = c(70,90), linetype=2, alpha=0.6) +
    annotate(geom='text',x=300, y=32, label=paste0('Best model mean pLDDT: ', model.dt[model == 'model_0',round(mean(pLDDT),1)]), size=5) +
    #annotate(geom='text',x=min.pos+40, y=92, label='excellent', color='black') +
    scale_color_manual(values=brewer.pal(5, 'Accent')) +
    scale_y_continuous(breaks=seq(0,140, by=10)) +
    scale_x_continuous(breaks=seq(0, 4579, by=200)) +
    ggtitle(paste0('Sacsin AF3 model0 pLDDT')) +
    ylab('pLDDT') +
    xlab('Amino acid residue') +
    theme_bw() +
    guides(alpha=guide_legend(title="pLDDT"))
 g 
  # I need to modify the anno dt to only extend as far as the region listed
  
q <- ggplot(sacsin.dt[region %in% c('UBL', 'SRR1', 'SRR2', 'SRR3','XPCB', 'DNAJ', 'HEPN'),], aes(xmin=start, xmax=end, y=molecule)) +
    geom_gene_arrow(fill = "white") +
    geom_subgene_arrow(data = sacsin.dt[region %in% c('UBL', 'SRR1', 'SRR2', 'SRR3','XPCB', 'DNAJ', 'HEPN'),], 
                       aes(xmin = start, xmax = end, y = molecule, fill = region, xsubmin = from, xsubmax = to), color="black", alpha=.7) +
    scale_fill_brewer(type='qual', palette=2) +
    theme_genes()


p <- ggplot(sacsin.dt[region %in% c('SIRPT1','SIRPT2','SIRPT3','Ubl','Hepn'),], aes(xmin=start, xmax=end, y=molecule)) +
    geom_gene_arrow(fill = "white") +
    geom_subgene_arrow(data = sacsin.dt[region %in% c('SIRPT1','SIRPT2','SIRPT3','Ubl','Hepn'),], 
                       aes(xmin = start, xmax = end, y = molecule, fill = region, xsubmin = from, xsubmax = to), color="black", alpha=.7) +
    scale_fill_brewer(type='qual', palette=1) +
    theme_genes()
  
patch <- g/q/p + plot_layout(heights=c(6,1,1))
patch
BackupAsPDF(patch, 'sacsin.AF3.colbyModel.plddt.linechart.Allgenetrack', dimensions = c(16,13))

```


```{r}

g <- ggplot(model.dt[model == 'model_0',], aes(x=auth_seq_id, y=pLDDT, color=pLDDT)) +
    geom_line(alpha=0.6, size=1) +
    geom_hline(yintercept = c(70,90), linetype=2, alpha=0.6) +
   # annotate(geom='text',x=300, y=32, label=paste0('Best model mean pLDDT: ', model.dt[model == 'model_0',round(mean(pLDDT),1)]), size=5) +
    #annotate(geom='text',x=min.pos+40, y=92, label='excellent', color='black') +
    scale_color_gradientn("plDDT", limits = c(0,100), colors = c(red = "#FE0012",
                                                      orange = "#FFA42B",
                                                      yellow = "#FFFD42","#FFFD42",
                                                      palegreen = "palegreen2",
                                                      blue = "lightblue","lightblue",#"#6097E8",
                                                      darkBlue = "#001DF9"),
                        values = c(0, .5, 0.7,0.75,0.8,0.85, 0.9, 1.0)) +
    scale_y_continuous(breaks=seq(0,140, by=10)) +
    scale_x_continuous(breaks=seq(0, 4579, by=200)) +
    ggtitle(paste0('Sacsin AF3 model0 pLDDT')) +
    ylab('pLDDT') +
    xlab('Amino acid residue') +
    theme_bw() +
    guides(alpha=guide_legend(title="pLDDT"))
 g 
  # I need to modify the anno dt to only extend as far as the region listed
  
q <- ggplot(sacsin.dt[region %in% c('UBL', 'SRR1', 'SRR2', 'SRR3','XPCB', 'DNAJ', 'HEPN'),], aes(xmin=start, xmax=end, y=molecule)) +
    geom_gene_arrow(fill = "white") +
    geom_subgene_arrow(data = sacsin.dt[region %in% c('UBL', 'SRR1', 'SRR2', 'SRR3','XPCB', 'DNAJ', 'HEPN'),], 
                       aes(xmin = start, xmax = end, y = molecule, fill = region, xsubmin = from, xsubmax = to), color="black", alpha=.7) +
    scale_fill_brewer(type='qual', palette=2) +
    theme_genes()


p <- ggplot(sacsin.dt[region %in% c('SIRPT1','SIRPT2','SIRPT3','Ubl','Hepn'),], aes(xmin=start, xmax=end, y=molecule)) +
    geom_gene_arrow(fill = "white") +
    geom_subgene_arrow(data = sacsin.dt[region %in% c('SIRPT1','SIRPT2','SIRPT3','Ubl','Hepn'),], 
                       aes(xmin = start, xmax = end, y = molecule, fill = region, xsubmin = from, xsubmax = to), color="black", alpha=.7) +
    scale_fill_brewer(type='qual', palette=1) +
    theme_genes()
  
patch <- g/q/p + plot_layout(heights=c(6,1,1))
patch
BackupAsPDF(patch, 'sacsin.AF3.model0.plddt.linechart.Allgenetrack', dimensions = c(16,13))

```

```{r}
lapply(subseqs, function(x){
  
  min.pos <- plDDT.dt[subsequence == x, min(aaResidue)]
  max.pos <- plDDT.dt[subsequence == x, max(aaResidue)]
  
  summary.dt <- plDDT.dt[subsequence == x, .(plddt, model, mean.plddt = mean(plddt)), by=aaResidue]
  
  g <- ggplot(summary.dt, aes(x=aaResidue, y=mean.plddt, color=mean.plddt)) +
    geom_line(alpha=0.6, size=1) +
    geom_hline(yintercept = c(70,90), linetype=2, alpha=0.6) +
    annotate(geom='text',x=min.pos+300, y=32, label=paste0('Best model mean pLDDT: ', summary.dt[model ==1,round(mean(plddt),1)]), size=3) +
    #annotate(geom='text',x=min.pos+40, y=92, label='excellent', color='black') +
    scale_color_gradientn("plDDT", limits = c(0,100), colors = c(red = "#FE0012",
                                                      orange = "#FFA42B",
                                                      yellow = "#FFFD42","#FFFD42",
                                                      palegreen = "palegreen2",
                                                      blue = "lightblue","lightblue",#"#6097E8",
                                                      darkBlue = "#001DF9"),
                        values = c(0, .5, 0.7,0.75,0.8,0.85, 0.9, 1.0)) +
    scale_y_continuous(breaks=seq(0,140, by=10)) +
    scale_x_continuous(breaks=seq(0, max.pos, by=100)) +
    ggtitle(paste0(x, ' Predicted LDDT residue scores')) +
    ylab('mean pLDDT') +
    xlab('Amino acid residue') +
    theme_bw() +
    guides(alpha=guide_legend(title="mean pLDDT"))
  
  # I need to modify the anno dt to only extend as far as the region listed
  anno.dt <- sacsin.dt[, `:=`(start = min.pos,
                               end = max.pos)]
  
  q <- ggplot(anno.dt[region %in% c('UBL', 'SRR1', 'SRR2', 'SRR3','XPCB', 'DNAJ', 'HEPN'),], aes(xmin=start, xmax=end, y=molecule)) +
    geom_gene_arrow(fill = "white") +
    geom_subgene_arrow(data = anno.dt[region %in% c('UBL', 'SRR1', 'SRR2', 'SRR3','XPCB', 'DNAJ', 'HEPN'),], aes(xmin = start, xmax = end, y = molecule, fill = region, xsubmin = from, xsubmax = to), color="black", alpha=.7) +
    scale_fill_brewer(type='qual', palette=2) +
    theme_genes()
  
  patch <- g/q + plot_layout(heights=c(5,1,1))
  BackupAsPDF(patch, paste0(x,'.mean.plddt.linechart.genetrack'), dimensions = c(14,10))
})

```
