---
title: "181223.tidyTables"
author: "Martin Gordon"
date: "2023-12-18"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 181223

Reformat the previously processed files and return the tables in the format requested

Read in the data shared by Danielle to use as an example
```{r}
library(ggplot2)
library(data.table)
library(magrittr)
library(ComplexHeatmap)
library(randomcoloR)
library(stringr)
library(OmnipathR)
library(circlize)
library(readxl)


#source BP utility scripts
source("/Users/martingordon/Documents/utils/bp_utils/ManageScriptData.R")
source ("/Users/martingordon/Documents/utils/bp_utils/MSstats_Helper_Functions.R")
source("/Users/martingordon/Documents/utils/bp_utils/MSstats_V4_Functions.R")
source ("/Users/martingordon/Documents/utils/bp_utils/UniprotIDMapping.R") #uniprot to gene id mapping
source ("/Users/martingordon/Documents/utils/bp_utils/enrichmentTestFunctions.R")
source ("/Users/martingordon/Documents/utils/bp_utils/linearModels.R")

# kinase enrichment analysis
source("/Users/martingordon/Documents/utils/bp_utils/KinaseActivityScores.R")

# mg utility functions
source("/Users/martingordon/Documents/utils/mg_utils/HelperFunctions.R")
```

Read in the data

```{r}
#this is just the quant info; maybe convert to wide and merge this with the pw comparisons that we have?
test <- fread('./data/MB03_PH_LibDIA_results.tsv')
test %>%  str()

ex.mss <- read_xlsx('/Users/martingordon/Documents/projects/112823_XZhong_DARPA2/data/20231011-PC1178-Roth_Phosphoproteome_Results.xlsx', sheet = 2, col_names = T, stri)
head(ex.mss)
getwd()
```
maybe he wants a combo of the intensities and the msstats results?

```{r}
getwd()  
mss <- fread('021223.ph.rereunWBatch_data/2023_12_02_mss.pwcomparisons.newbatch.raw.csv')

dp.out <- readRDS('~/Documents/projects/112823_XZhong_DARPA2/DE.ph.LMredo.noBatch_data/2023_11_30_ph.mss.dataproc')
p.quant<- setDT(dp.out$ProteinLevelData)

#also read in the raw data as may need some additional info from here
spec <- fread("./output/MSstatsInput.PH.csv")

dcast(spec, paste0(ProteinName, '_', PrecursorCharge)~paste0(Condition, '_', BioReplicate), value.var = c('PeptideSequence')) %>%  View()

spec[ProteinName == 'O88935_S9' & Condition == 'M100' & BioReplicate == 1,]
```

```{r}
# convert p.quant to wide format and get the log intensities for eq
p.quant.wide <-  dcast(p.quant, Protein~GROUP+SUBJECT, value.var = c('LogIntensities'))

# read in the raw msstats file
mss <- fread('021223.ph.rereunWBatch_data/2023_12_02_mss.pwcomparisons.newbatch.raw.csv')

mss.wide <- dcast(mss, Protein~Label, value.var = c('log2FC', 'pvalue', 'adj.pvalue')) 

mss.wide
colnames(p.quant.wide)[-1] <-  paste0('LogIntensities_')

colnames(p.quant.wide)[-1] <- paste0('LogIntensities_', colnames(p.quant.wide)[-1])
merge.table <- merge(x=mss.wide, y=p.quant.wide, on='Protein')

merge.table[, gene_site := multiUniprotSites2multiGeneSites(Protein, species='MOUSE')]
merge.table[, protein_site :=  Protein]
merge.table[, Protein := gsub('_[STY]{1}[0-9]+', '', Protein)]
merge.table[, Gene := multiUniprots2multiGenes(Protein, species='MOUSE')]

merge.table[, site := sapply(str_extract_all(Protein, '_[STY]{1}[0-9]+'), paste, collapse=',')]
merge.table[, site := sapply(str_extract_all(protein_site, '_[STY]{1}[0-9]+'), paste, collapse=',')]
merge.table[, site := gsub('_','', site)]

# count the number of phospho sites
merge.table[, ph.multiplicity := str_count(site, '[STY]') ]

merge.table %>%  View()

merge.table %>%  colnames()


col.order <- c("Protein", 'Gene', 'protein_site', 'ph.multiplicity', 
               "log2FC_M100-Veh", "log2FC_M100_Nbon-M100", "log2FC_M100_Nbon-Nbon", "log2FC_M100_Nbon-Veh","log2FC_Nbon-Veh",
               "pvalue_M100-Veh", "pvalue_M100_Nbon-M100", "pvalue_M100_Nbon-Nbon",  "pvalue_M100_Nbon-Veh" ,  "pvalue_Nbon-Veh",
               "adj.pvalue_M100-Veh", "adj.pvalue_M100_Nbon-M100", "adj.pvalue_M100_Nbon-Nbon", "adj.pvalue_M100_Nbon-Veh", "adj.pvalue_Nbon-Veh",
               "LogIntensities_M100_1", "LogIntensities_M100_2", "LogIntensities_M100_3", "LogIntensities_M100_4",
               "LogIntensities_M100_Nbon_1", "LogIntensities_M100_Nbon_2", "LogIntensities_M100_Nbon_3", "LogIntensities_M100_Nbon_4",
               "LogIntensities_Nbon_1","LogIntensities_Nbon_2","LogIntensities_Nbon_3","LogIntensities_Nbon_4",
               "LogIntensities_Veh_1","LogIntensities_Veh_2","LogIntensities_Veh_3","LogIntensities_Veh_4")

merge.table[, ..col.order]
 sapply(str_extract_all(VARIANT.ANNO, aa.pattern), paste, collapse = ',')]

merge.table[, site := str_extract_all('_[STY]{1}',Protein)]
merge.table

str_extr

p.quant[ Protein %like% 'A0A087WPF7',]
merge.table[Protein == 'A0A087WPF7']

p.quant
str_extract_all()

fwrite(merge.table, './output/Phres.merge.fc.abu.wide.tsv', sep = '\t')
fread('./output/Phres.merge.fc.abu.wide.csv')

merge.table %>% head()
```

as is share the table with Bryan in wide format, ex

