---
title: "T_SNE_netprop"
author: "Martin Gordon"
date: "2023-06-29"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library("RColorBrewer") 
library(data.table)
library(stringr) #str_extract function 
library(magrittr)
library(ggplot2)
library (ComplexHeatmap)
library(RcppCNPy) #R/W support to import numpy objects into R
library(VennDiagram)
library(pbapply) #apply functions with progress bar
library(eulerr)

# load bens util scripts for the analysis
source("../../utils/bp_utils/ManageScriptData.R")
source("../../utils/bp_utils/LocatePeptidePtmInProteins.R")
source ("../../utils/bp_utils/UniprotIDMapping.R") #uniprot to gene id mapping
source ("../../utils/bp_utils/enrichmentTestFunctions.R")

# netprop scripts
source ("../../utils/bp_utils/STRING_db_utils.R")
source ("../../utils/bp_utils/NetworkPropagation.R")


# Attach the `umap` library
library(umap)

library(cluster) #k-mediod clustering
library(factoextra) #estimate optimal k
```


## Generate a t-sne plot of the output from the netprop results


For now use the original run data as the repeat results were quite poor

```{r}
rna.ab.allNPdt <- fread("./output/AllNetworkPropagation_RNA_STRING_Interactions.rs0.3.csv")
rna.ab.allNPdt[, omics :='rna']
prot.ab.allNPdt <- fread("./output/AllNetworkPropagation_PROT_AB_STRING_Interactions.rs0.3.csv")
prot.ab.allNPdt[, omics := 'p.ab']
prot.ph.allNPdt <-fread("./output/AllNetworkPropagation_PROT_PH_STRING_Interactions.rs0.3.csv")
prot.ph.allNPdt[, omics := 'p.ph']

```

Combine the output of the different net.prop results

```{r}
np.out.list <- list('rna_ab' = rna.ab.allNPdt,
                      'prot_ab' = prot.ab.allNPdt,
                      'prot_ph' = prot.ph.allNPdt)


np.out.comb  <- do.call(rbind, np.out.list) #colbind
fwrite(np.out.comb, "./output/AllNetworkPropagation_AllOmics_STRING_Interactions.rs0.3.csv")

#np.out.comb.list <- split(np.out.comb, np.out.comb$Label)

# get gene names
np.out.comb[, symbol :=  translateString2Gene(gene, species= "HUMAN")]

```

Perform Enrichment Analysis on the 3 Different Groups & the Sig Genes

```{r}
gmt <- loadGmtFromBioconductor(ontology = "ALL", keyType = "SYMBOL")

enrich.dt <- enricherOnGroups(groupTable = np.out.comb[adj.pvalue < 0.05], geneColumn = "symbol", groupColumns = c("omics", "Label"),term2gene.gmt = gmt, universe = unique(np.out.comb$symbol), numProcessors = 8 )

simp <- simplifyEnrichBySimilarUniverseMembership(enrich.dt, gmt, groupColumn = "omics.Label")

pdf('./output/np-heatmap-output.pdf')

hm <- enrichHeatmapBestPerGroup(simplifiedEnrichTable = simp[[1]], fullEnrichTable = simp[[2]],
                                      pipeLineList = NULL,
                                      groupColumn="omics.Label", topN = 15, title="", cols = NULL, 
                                      negCols = NULL, reduceRedundantsAcrossGroups=TRUE, max_pAdjust = 0.01, minCount = 1,
                                      annotatePossibleMatches = TRUE,  row_names_gp = gpar(fontsize = 10),
                                      upperThreshold  = 10, column_names_gp = gpar(fontsize = 8),
                                      pvalColumn = "p.adjust")

draw(hm$hmList)
dev.off()
```
Enrichment analysis comparing the conditions:

```{r}
gmt <- loadGmtFromBioconductor(ontology = "ALL", keyType = "SYMBOL")

enrich.dt <- enricherOnGroups(groupTable = np.out.comb[adj.pvalue < 0.05], geneColumn = "symbol", groupColumns = c( "Label"),term2gene.gmt = gmt, universe = unique(np.out.comb$symbol), numProcessors = 8 )

simp <- simplifyEnrichBySimilarUniverseMembership(enrich.dt, gmt, groupColumn = "Label")

pdf('./output/np-heatmap-condition.pdf')

hm <- enrichHeatmapBestPerGroup(simplifiedEnrichTable = simp[[1]], fullEnrichTable = simp[[2]],
                                      pipeLineList = NULL,
                                      groupColumn="Label", topN = 30, title="", cols = NULL, 
                                      negCols = NULL, reduceRedundantsAcrossGroups=TRUE, max_pAdjust = 0.01, minCount = 1,
                                      annotatePossibleMatches = TRUE,  row_names_gp = gpar(fontsize = 10),
                                      upperThreshold  = 30, column_names_gp = gpar(fontsize = 8),
                                      pvalColumn = "p.adjust")

draw(hm$hmList)
dev.off()
```


```{r}
# read in the clean omics files
rna.ab.clean <- fread('./output/rna_ab.clean.csv.gz')
rna.ab.clean[, omics :='rna']
prot.ab.clean <- fread('./output/prot_ab.clean.csv.gz')
prot.ab.clean[, omics := 'p.ab']
prot.ph.clean <- fread('./output/prot_ph.clean.csv.gz')
prot.ph.clean[, omics := 'p.ph']

omics.clean <- list(rna_ab = rna.ab.clean,
                    prot_ab = prot.ab.clean,
                    prot_ph = prot.ph.clean)
```

Look at the number of genes in all 3 that are significant

lets also view the overlap in the number of differentially upregulated features between the three datasets

Collapse the 3 lists to one data.table, and split on contrast
```{r}
merge.omics <- rbindlist(omics.clean, fill = T,idcol = T)

#Add sig column for venn diagram visualisation of differentially expressed features

merge.omics[, sig := 'not']
merge.omics[ fdr < 0.05 & abs(log2FC) > log2(1.5), sig := ifelse(log2FC < 0, "down", "up")]
merge.omics[, sig := factor(sig, levels = c("up", "not", "down"))]

merge.omics.list <- split(merge.omics, merge.omics$Label)
```

# redundancy in prot_ph
Break down of sig results among the three datasets; most sig results in phosphoproteomics?
# ph data: 1111 unique sig in cond 1, 272 unique sig in condition 2

```{r}
lapply(seq_along(merge.omics.list), function(x,n,i){
  x[[i]][sig != 'not' & .id == 'prot_ph', .N, by=.(gene)][order(-N)]
}, x=merge.omics.list, n=names(merge.omics.list))
``` 

Removed duplicates if over level of significance
#in total 1645 genes sig in label 1, 496 in label 2
```{r}

sig.genes.list <- lapply(seq_along(merge.omics.list), function(x,n,i){
  #cal geom.mean 
  x[[i]][, gm.lfc.fdr := sqrt(abs(log2FC) * fdr)] %>% 
    .[.id == 'prot_ph', gene := multiUniprots2multiGenes(uniprot, simplify = F)] %>% #uniprot col is the clean version in prot_ph dataser
    # only take the row with the max geom.mean value in the ph group
    .[sig != 'not', .SD[which.max(gm.lfc.fdr)], by = .(.id,gene)]
}, x=merge.omics.list, n=names(merge.omics.list))

#combine on label for enrichment
sig.genes <- do.call(rbind, sig.genes.list)
```

```{r}
enrich.dt <- enricherOnGroups(groupTable = sig.genes, geneColumn = "gene", groupColumns = c("omics", "Label"),term2gene.gmt = gmt, universe = unique(merge.omics$gene), numProcessors = 8 )

simp <- simplifyEnrichBySimilarUniverseMembership(enrich.dt, gmt, groupColumn = "omics.Label")

pdf('./output/exp-heatmap-output.pdf')

hm <- enrichHeatmapBestPerGroup(simplifiedEnrichTable = simp[[1]], fullEnrichTable = simp[[2]],
                                      pipeLineList = NULL,
                                      groupColumn="omics.Label", topN = 15, title="", cols = NULL, 
                                      negCols = NULL, reduceRedundantsAcrossGroups=TRUE, max_pAdjust = 0.01, minCount = 1,
                                      annotatePossibleMatches = TRUE,  row_names_gp = gpar(fontsize = 10),
                                      upperThreshold  = 10, column_names_gp = gpar(fontsize = 8),
                                      pvalColumn = "p.adjust")

draw(hm$hmList)
dev.off()
```
Enrichment plot of the sig things in both conditions

```{r}
enrich.dt <- enricherOnGroups(groupTable = sig.genes, geneColumn = "gene", groupColumns = c("Label"),term2gene.gmt = gmt, universe = unique(merge.omics$gene), numProcessors = 8 )

simp <- simplifyEnrichBySimilarUniverseMembership(enrich.dt, gmt, groupColumn = "Label")

pdf('./output/exp-heatmap-condition.pdf')

hm <- enrichHeatmapBestPerGroup(simplifiedEnrichTable = simp[[1]], fullEnrichTable = simp[[2]],
                                      pipeLineList = NULL,
                                      groupColumn="Label", topN = 30, title="", cols = NULL, 
                                      negCols = NULL, reduceRedundantsAcrossGroups=TRUE, max_pAdjust = 0.01, minCount = 1,
                                      annotatePossibleMatches = TRUE,  row_names_gp = gpar(fontsize = 10),
                                      upperThreshold  = 10, column_names_gp = gpar(fontsize = 8),
                                      pvalColumn = "p.adjust")

draw(hm$hmList)
dev.off()
```
Create a t-SNE plot

For input we will use all significant genes in the 3 datasets
We want 'informative' features; those with i) significant change in more than one dataset or ii) significant change in one dataset 

Take the significant genes from each NP set

Read in the original netprop results

```{r}
#reload 
rna.ab.allNPdt <- fread("./output/AllNetworkPropagation_RNA_STRING_Interactions.rs0.3.csv")
prot.ab.allNPdt <- fread("./output/AllNetworkPropagation_PROT_AB_STRING_Interactions.rs0.3.csv")
prot.ph.allNPdt <-fread("./output/AllNetworkPropagation_PROT_PH_STRING_Interactions.rs0.3.csv")

np.out.list <- list('rna.ab_np' = rna.ab.allNPdt,
                      'prot.ab_np' = prot.ab.allNPdt,
                      'prot.ph_np' = prot.ph.allNPdt)
```
Extract results with an fdr < 0.1. These genes will be input for T-SNE plot

Plot the overlap of the significant NP output

```{r}

sig.np.omics.genes <- lapply(np.out.list, function(x){
  x[ adj.pvalue < 0.1 & abs(z) > 1.96, .(gene, symbol = translateString2Gene(gene, species='HUMAN'),Label)] 
})

# extrac the genes for each comparison
sig.np.EV.WT <- lapply(sig.np.omics.genes, function(x){
  x[ Label == 'EV_IFNg-vs-WT_IFNg',symbol] 
})

sig.np.WT.ND <- lapply(sig.np.omics.genes, function(x){
  x[ Label == 'WT_IFNg-vs-ND_IFNg', symbol] 
})


p <- plot(eulerr::euler(sig.np.WT.ND),
              quantities = TRUE, main = c('WT_IFNg-vs-ND_IFNg NP features (adj.pval < 0.1)', cex=1))

p

png(ScriptAndDatedFileName("WT_IFNg-vs-ND_IFNg.netprop.sigfeatures.png"), height = 5, width = 6, units = "in", res = 200)
p
dev.off()
p



p <- plot(eulerr::euler(sig.np.EV.WT),
              quantities = TRUE, main = c('EV_IFNg-vs-WT_IFNg NP features (adj.pval < 0.1)', cex=1))

p

png(ScriptAndDatedFileName("EV_IFNg-vs-WT_IFNg.netprop.sigfeatures.png"), height = 5, width = 6, units = "in", res = 200)
p
dev.off()
p
```

We can see in general that most significant hits do not overlap, with only one feature observed in one condition in all 3 omics datasets

Lets do the same for the experimental data;

```{r}
# read in the clean omics files
rna.ab.clean <- fread('./output/rna_ab.clean.csv.gz')
rna.ab.clean[, adj.pvalue := p.adjust(pvalue, method = 'BH')]
prot.ab.clean <- fread('./output/prot_ab.clean.csv.gz')
prot.ph.clean <- fread('./output/prot_ph.clean.csv.gz')

omics.clean <- list(rna.ab_exp = rna.ab.clean,
                    prot.ab_exp = prot.ab.clean,
                    prot.ph_exp = prot.ph.clean)
```

For significant, we will consider log2 FC > 0.58 & adj.pvalue < 0.05


```{r}

sig.exp.omics.genes <- lapply(omics.clean, function(x){
  x[abs(log2FC) > log2(1.5) & adj.pvalue < 0.05,]
})

```

Simplify the phospho output; just take a single gene for each phospho output; we will take largest log2FC

```{r}
# this may be incorrect; use geometric mean 
sig.exp.omics.genes[['prot.ph_exp']] <-  sig.exp.omics.genes[['prot.ph_exp']][!is.na(log2FC) & !is.na(adj.pvalue), .(log2FC = max(log2FC, na.rm = T)), by = .(Label, uniprot)]
sig.exp.omics.genes[['prot.ph_exp']][,gene := translateUniprot2GeneName(uniprot)]

```

```{r}
# extrac the genes for each comparison
sig.exp.EV.WT <- lapply(sig.exp.omics.genes, function(x){
  x[ Label == 'EV_IFNg-vs-WT_IFNg',gene] %>% 
    unique()
})

sig.exp.WT.ND <- lapply(sig.exp.omics.genes, function(x){
  x[ Label == 'WT_IFNg-vs-ND_IFNg', gene] %>% 
    unique()
})


p <- plot(eulerr::euler(sig.exp.WT.ND),
              quantities = TRUE, main = c('WT_IFNg-vs-ND_IFNg EXP features (adj.pval < 0.05 & log2FC < 0.58)', cex=1))

p

png(ScriptAndDatedFileName("WT_IFNg-vs-ND_IFNg.experimental.sigfeatures.png"), height = 5, width = 6, units = "in", res = 200)
p
dev.off()
p



p <- plot(eulerr::euler(sig.exp.EV.WT),
              quantities = TRUE, main = c('EV_IFNg-vs-WT_IFNg EXP features (adj.pval < 0.05 & log2FC < 0.58)', cex=1))

p

png(ScriptAndDatedFileName("EV_IFNg-vs-WT_IFNg.experimental.sigfeatures.png"), height = 5, width = 6, units = "in", res = 200)
p
dev.off()
p
```
Combine the two lists and plot the results

```{r}
all.sig.EV.WT <- c(sig.exp.EV.WT, sig.np.EV.WT)

p <- plot(eulerr::euler(all.sig.EV.WT),
              quantities = TRUE, main = c('EV_IFNg-vs-WT_IFNg All Sig features', cex=1))

png(ScriptAndDatedFileName("EV_IFNg-vs-WT_IFNg.all.sigfeatures.png"), height = 5, width = 6, units = "in", res = 200)
p
dev.off()
p


all.sig.WT.ND <- c(sig.exp.WT.ND, sig.np.WT.ND)

p <- plot(eulerr::euler(all.sig.WT.ND),
              quantities = TRUE, main = c('WT_IFNg-vs-ND_IFNg All Sig features', cex=1))

p

png(ScriptAndDatedFileName("WT_IFNg-vs-ND_IFNg.all.sigfeatures.png"), height = 5, width = 6, units = "in", res = 200)
p
dev.off()
p
```
Modularise the network of NP significant genes

```{r}
#for looping
contrasts.OI <- c('EV_IFNg-vs-WT_IFNg', 'WT_IFNg-vs-ND_IFNg')
combinations <- expand.grid(X = sig.np.omics.genes, Y = contrasts.OI)


#combinethe netprop sig results & take the unique gene list
sig.np.df <- rbindlist(sig.np.omics.genes, idcol = 'omics')

#sigfeatures.np.EV.WT <- unique(sig.np.EV.WT.df$symbol)


sig.np.WT.ND.df <- rbindlist(sig.np.omics.genes, idcol = 'omics') %>% 
  .[Label == 'WT_IFNg-vs-ND_IFNg',]

sigfeatures.np.WT.ND <- unique(sig.np.WT.ND.df$symbol)



# get string subnetwork for each contrast
string.net.list <- lapply(contrasts.OI, function(i){
  
  sigfeatures <- sig.np.df[Label == i, unique(symbol)]
  stringMapping <- GetStringIDMapping(sigfeatures, stringAliasFile='./data/9606.protein.aliases.v11.5.txt.gz')
  
  stringsOI <- c(setdiff(sigfeatures, stringMapping$alias), stringMapping$string)
  stringEdges <- GetStringSubNetwork(stringsOI,  oneHopConnections = TRUE, stringFile = './data/9606.protein.links.detailed.v11.5.txt.gz' ) #includes 1 hop connectors

  })
# add names to the dataset
names(string.net.list) <- contrasts.OI



lapply(seq_along(string.net.list), function(x,n,i){
  
  fwrite(x[[i]], paste0("./data/", n[[i]], "StringSubNetwork.1hop.txt"), sep = "\t", col.names=FALSE)
  },x=string.net.list, n=names(string.net.list))

```

### DSD in external python script

```{r, eval=FALSE}
# run this in CL
#conda activate python2.7
# -c confidence values in ppi as edge weights
# -s 20 steps in random walk

#for i in EV_IFNg-vs-WT_IFNg WT_IFNg-vs-ND_IFNg; do echo " running DSD on $i"; python ./scripts/capDSD-src/DSD.py ./data/${i}StringSubNetwork.1hop.txt ./data/${i}StringSubNetwork.1hop; done

```


Lets try run with one dataframe

```{r}
dsd.wt.nd <-fread('./data/WT_IFNg-vs-ND_IFNgStringSubNetwork.1hop.dsd')
dsd.mat <- as.matrix(dsd.wt.nd, rownames='V1')

sigfeatures.np.WT.ND
stringMapping <- GetStringIDMapping(sigfeatures.np.WT.ND, stringAliasFile='./data/9606.protein.aliases.v11.5.txt.gz')
stringsOI <- c(setdiff(sigfeatures.np.WT.ND, stringMapping$alias), stringMapping$string)

attr(dsd.mat, 'Size')

# remove the 1 hop connectors
stringsInMat <- intersect(rownames(dsd.mat), c(stringsOI))
dsd.mat <- dsd.mat[stringsInMat, stringsInMat]

dim(dsd.mat)

View(hclust)
 as.integer(attr(dsd.dist, "Size"))
dsd.dist <- as.dist (dsd.mat)
# ddr <- hclust (dsd.dist) %>% as.dendrogram()
# Heatmap(dsd.mat, show_row_names = FALSE, show_column_names = FALSE, cluster_rows = ddr, cluster_columns = ddr)
# ddr <- hclust (dsd.dist, method= "ward.D") %>% as.dendrogram()
# Heatmap(dsd.mat, show_row_names = FALSE, show_column_names = FALSE, cluster_rows = ddr, cluster_columns = ddr)
# ddr <- hclust (dsd.dist, method= "ward.D2") %>% as.dendrogram()
# Heatmap(dsd.mat, show_row_names = FALSE, show_column_names = FALSE, cluster_rows = ddr, cluster_columns = ddr)
ddr <- hclust (dsd.dist, method= "average") |> as.dendrogram()
Heatmap(dsd.mat, show_row_names = T, show_column_names = FALSE, cluster_rows = ddr, cluster_columns = ddr)
```

```{r, fig.width = 12, fig.height = 12}
dsd.ev.wt <- fread ("./data/EV_IFNg-vs-WT_IFNgStringSubNetwork.1hop.dsd")
dsd.wt.nd <-fread('./data/WT_IFNg-vs-ND_IFNgStringSubNetwork.1hop.dsd')

dsd.list <- list("EV_IFNg-vs-WT_IFNg" = dsd.ev.wt,
                 "WT_IFNg-vs-ND_IFNg" = dsd.wt.nd)

#convert to matrix
dsd.list <- lapply(dsd.list, function(x){x <- as.matrix(x, rownames='V1')})

# create subnet distance mat 
string.dist.list <- lapply(contrasts.OI, function(x,y,i){
  
  sigfeatures <- y[Label == i, unique(symbol)]
  stringMapping <- GetStringIDMapping(sigfeatures, stringAliasFile='./data/9606.protein.aliases.v11.5.txt.gz') # convert to string
  # get interesting strings
  stringsOI <- c(setdiff(sigfeatures, stringMapping$alias), stringMapping$string) #return elements unique in sigfeatures not in str$aliad
  
  #remove the 1 hop connectors
  stringsInMat <- intersect(rownames(x[[i]]), c(stringsOI))
  submat <- x[[i]][stringsInMat, stringsInMat]
  
  },x=dsd.list, y=sig.np.df)

names(string.dist.list) <- contrasts.OI


lapply(seq_along(string.dist.list), function(x,n,i){
  
  #get distance matrix for clustering
  dist.mat <- as.dist(x[[i]])
  ddr <- hclust(dist.mat, method= "average") %>% 
    as.dendrogram()

  hm <- Heatmap(x[[i]], show_row_names = FALSE, show_column_names = FALSE, cluster_rows = ddr, cluster_columns = ddr)
  BackupAsPDF(draw(hm),paste0(n[[i]],'dissimilarity-matrix.pdf'))

  
}, x=string.dist.list, n=names(string.dist.list))

```
### tsne based on DSD

For visualizing the modularization later

```{r, fig.width = 6, fig.height =  6}

tsne.list <- lapply(seq_along(string.dist.list), function(x,n,i){

 dist.mat <- as.dist(x[[i]])
 t <- Rtsne::Rtsne(dist.mat, is.distance = TRUE, theta = 0.0 ) # exact tsne when theta = 0.0

 tsne.dt <- as.data.table(t$Y)
 tsne.dt$string <- attr(dist.mat, "Labels")
 return(tsne.dt)

}, x= string.dist.list, n=names(string.dist.list))


names(tsne.list) <- contrasts.OI

lapply(tsne.list, function(x){ 
  ggplot(x, aes(x=V1, y=V2)) + geom_point()})
```
# modularization

Going to use k-medoid clustering (more robust than k-means)


First use the `factoextra` package to estimate otpimal number of clusters
Weighted sum of squares suggests that 5~6 clusters works for both

```{r}

names(string.dist.list)

lapply(string.dist.list, function(x){
  #fviz_nbclust(x, kmeans,method='wss', k.max = 14) #5-7
})

names(string.dist.list)
```
Plot the T-SNE modules
```{r, fig.width = 6, fig.height =  6}

# cluster.dt <- defineModulesFromDistance("/Users/ben/UCSF/kroganlab/BenPolacco/roche/2023_06_02_RSV_NetProp_HeatDiffusion_data/2023_06_05_StringSubNetwork.Preys.1hop.dsd",geneNamesOI = stringsOI, deepSplit = 0.8)
# tsne.dt[cluster.dt, c("clusterID", "cluster") := .(i.clusterID, i.cluster), on = c(string = "gene")]


lapply(contrasts.OI, function(x,y,i){
  
  pam.out <- cluster::pam(x[[i]],diss = TRUE, k = 7)
  y[[i]][names(pam.out$clustering), clusterID := pam.out$clustering, on = "string"]

  y[[i]][, cluster := sprintf("clust.%02d", clusterID)]
  y[[i]][, c("x", "y") := .(V1, V2)]

  y[[i]][, .N, by=cluster]
  
 p <- ggplot(y[[i]], aes (x = x, y = y,)) +
  #geom_segment(data = edgeView.dt, aes(xend = xend, yend = yend),  alpha = 0.2, linewidth = 0.1) +
 geom_point(aes( color = cluster)) +
 #scale_size_manual(values = c(bait = 4, prey = 0.5)) +
 #scale_shape_manual(values = c(bait = 23, prey = 1)) +
 coord_fixed() + theme_void()
p

}, x=string.dist.list, y=tsne.list)

```

## enrich and label the modules
```{r}
gmt <- loadGmtFromBioconductor(ontology = "ALL", keyType = "SYMBOL")


lapply(seq_along(tsne.list), function(x,n,i){
  x[[i]][, gene := translateString2Gene(string, species = "HUMAN")]
  geneToCluster <- x[[i]][,.(gene = gene, cluster = cluster)]
  
  #label each of the modules by enrichment results
  enrichNames.out <- labelModulesByEnrichment(cluster.dt = geneToCluster, gmt = gmt, numProcessors = 8)
  
  # choose a naming scheme
 clusterNames.dt <- enrichNames.out$clusterNames.dt[, .(cluster, name = termScore.name, go = termScore.go)]
 enrich.dt <- enrichNames.out$enrich.dt
 
 x[[i]][clusterNames.dt, c("name", "go" ) := .( i.name, i.go), on = "cluster"]
 x[[i]][gmt, nameMatch := TRUE, on = c(gene = "gene", go = "ont") ]
 x[[i]][nameMatch == TRUE, nameMatchCluster := cluster]
  
 
 p <- ggplot(x[[i]], aes (x = x, y = y,)) +
  #geom_segment(data = edgeView.dt, aes(xend = xend, yend = yend),  alpha = 0.2, linewidth = 0.1) +
  geom_point(aes( color = name  )) +
   ggtitle(paste0(n[[i]], ' feature (adj.pval < 0.1) enrichment' )) +
  #scale_size_manual(values = c(bait = 4, prey = 0.5)) +
  #scale_shape_manual(values = c(bait = 23, prey = 1)) +
  coord_fixed() + theme_void()
p

#BackupAsPDF(p, paste0(n[[i]],'tsne.enrichment.pdf'))

}, x=tsne.list, n=names(tsne.list))


#on plot: gene ratio 19/23 # ratio of the number of genes within a given gene set that are present in the input gene list
#bg ratio 22/215 #null set what you would expect by chance. 22 here respents count in the universe/background set
```

Get the enrichment tables

```{r}
lapply(seq_along(tsne.list), function(x,n,i){
  x[[i]][, gene := translateString2Gene(string, species = "HUMAN")]
  geneToCluster <- x[[i]][,.(gene = gene, cluster = cluster)]
  
  #label each of the modules by enrichment results
  enrichNames.out <- labelModulesByEnrichment(cluster.dt = geneToCluster, gmt = gmt, numProcessors = 8)
  # choose a naming scheme
  clusterNames.dt <- enrichNames.out$clusterNames.dt[, .(cluster, name = termScore.name, go = termScore.go)]
 # print(clusterNames.dt)
 fwrite(clusterNames.dt, paste0('./output/',n[[i]], '-t-SNE.cluster.go.enrichment.csv.gz'))


}, x=tsne.list, n=names(tsne.list))
```


Can we label where the datapoint came from?
Map to string - test one example




```{r}
sig.np.df

contrasts.OI

test.df <- sig.np.df[Label == 'WT_IFNg-vs-ND_IFNg']
test.tsne.df <- tsne.list[['WT_IFNg-vs-ND_IFNg']]

stringMapping
test.df[stringMapping, string := i.string, on = c(symbol = 'alias')] #assigns values from string col in stringMapping to a var called string in test.df
test.df[is.na(string), string := symbol]


  sigfeatures <- y[Label == i, unique(symbol)]
  stringMapping <- GetStringIDMapping(sigfeatures, stringAliasFile='./data/9606.protein.aliases.v11.5.txt.gz') # convert to string

  y[[i]][stringMapping, string := i.string, on = c(symbol = "alias")]
  y[[i]][is.na(string), string := V1]

omicsMapped <- test.df[, paste(sort(omics), collapse= "_"), by = string]


test <- test.tsne.df[omicsMapped, omics := i.V1, on = "string"]
test
omicsMapped

test.tsne.df


p <- ggplot(test.tsne.df, aes (x = x, y = y,)) +
  #geom_segment(data = edgeView.dt, aes(xend = xend, yend = yend),  alpha = 0.2, linewidth = 0.1) +
  geom_jitter(aes( color = name , shape = omics), height = 1, width = 1, ) +
  coord_fixed() + theme_void()
p
```


test.df

stringMapping
#
#combinethe netprop sig results & take the unique gene list
#sig.np.df <- rbindlist(sig.np.omics.genes, idcol = 'omics')
sig.np.df

```{r}
# looping through t-sne list and the sig-gene list at same time to find sig genes
# match gene in df with string in t-sne; just need to handle mutlimatches
lapply(contrasts.OI, function(x,y,i){

 y[Label == i, string := gene] #create string lab to match tsne-list
 omicsMapped <- y[Label == i,][, paste(sort(omics), collapse= "_"), by = string] #collapse together omics found in multiple genes
  
 x[[i]][omicsMapped, omics := i.V1, on = "string"]
 
p <- ggplot(x[[i]], aes (x = x, y = y,)) +
  geom_jitter(aes( color = name , shape = omics), height = 1, width = 1, ) +
  ggtitle(paste0(i)) +
  scale_shape_manual(values = c(rna.ab_np  = 0, prot.ph_np = 2, prot.ab_np = 3, prot.ab_np_rna.ab_np=16, prot.ph_np_rna.ab_np=17, prot.ab_np_prot.ph_np_rna.ab_np=15)) +
  coord_fixed() + theme_void() + theme(
    #plot.title = element_text(size=10, hjust=-0.5, vjust=4),
    legend.title = element_text(size = 6), 
    legend.text = element_text(size = 6))
p
BackupAsPDF(p, paste0(i,'tsne.omics.enrichment.pdf'))

  
  }, x=tsne.list, y=sig.np.df)
```




p <- ggplot(tsne.dt, aes (x = x, y = y,)) +
  #geom_segment(data = edgeView.dt, aes(xend = xend, yend = yend),  alpha = 0.2, linewidth = 0.1) +
  geom_jitter(aes( color = name , shape = virus, alpha = virus), height = 1, width = 1, ) +
  #ggrepel::geom_text_repel(data = tsne.dt[virus == "PIV_RSV"], mapping = aes(label = gene), size = 2, color = "black", show.legend = FALSE) +
  #scale_size_manual(values = c(bait = 4, prey = 0.5)) +
  #scale_shape_manual(values = c(bait = 23, prey = 1)) +
  scale_shape_manual(values = c(PIV  = 0, RSV = 2, PIV_RSV = 19)) +
  scale_alpha_manual(values = c(PIV = 0.5, RSV = 0.5, PIV_RSV = 0.8)) + 
  coord_fixed() + theme_void()
p
```


View(sig.np.df)

sets.dt[stringMapping, string := i.string, on = c(V1 = "alias")]
sets.dt[is.na(string), string := V1]

virusMapped <- sets.dt[, paste(sort(virus), collapse= "_"), by = string]

tsne.dt[virusMapped, virus := i.V1, on = "string"]
```




tsne.dt[, gene := translateString2Gene(string, species = "HUMAN")]
#uniprots <- stringMapping[cluster.dt$gene, alias, on = "string"]
geneToCluster <- tsne.dt[,.(gene = gene, cluster = cluster)]

#source ("../../bp_utils/enrichmentTestFunctions.R")
# we are in string/uniprot land. Load a gmt that we can map to string
#gmt <- loadGmtFromBioconductor(ontology = "ALL", keyType = "UNIPROT")
#gmt.string <- gmt[stringMapping, .(term = ont, gene = i.string) , on = c(gene = "alias")][!is.na(term)]

enrichNames.out <- labelModulesByEnrichment(cluster.dt = geneToCluster, gmt = gmt, numProcessors = 8)

# choose a naming scheme
clusterNames.dt <- enrichNames.out$clusterNames.dt[, .(cluster, name = termScore.name, go = termScore.go)]
enrich.dt <- enrichNames.out$enrich.dt
```

```{r}

tsne.dt[clusterNames.dt, c("name", "go" ) := .( i.name, i.go), on = "cluster"]

tsne.dt[gmt, nameMatch := TRUE, on = c(gene = "gene", go = "ont") ]

tsne.dt[nameMatch == TRUE, nameMatchCluster := cluster]

```

```{r, fig.width = 10, fig.height = 6}
p <- ggplot(tsne.dt, aes (x = x, y = y,)) +
  #geom_segment(data = edgeView.dt, aes(xend = xend, yend = yend),  alpha = 0.2, linewidth = 0.1) +
  geom_point(aes( color = name  )) +
  #scale_size_manual(values = c(bait = 4, prey = 0.5)) +
  #scale_shape_manual(values = c(bait = 23, prey = 1)) +
  coord_fixed() + theme_void()
p
```



```{r}
sets.dt <- rbindlist(lapply(bothVirus, data.table), idcol = "virus")

sets.dt[stringMapping, string := i.string, on = c(V1 = "alias")]
sets.dt[is.na(string), string := V1]

virusMapped <- sets.dt[, paste(sort(virus), collapse= "_"), by = string]

tsne.dt[virusMapped, virus := i.V1, on = "string"]
```




```{r, fig.width = 8, fig.height = 4}
p <- ggplot(tsne.dt, aes (x = x, y = y,)) +
  #geom_segment(data = edgeView.dt, aes(xend = xend, yend = yend),  alpha = 0.2, linewidth = 0.1) +
  geom_jitter(aes( color = name , shape = virus, alpha = virus), height = 1, width = 1, ) +
  #ggrepel::geom_text_repel(data = tsne.dt[virus == "PIV_RSV"], mapping = aes(label = gene), size = 2, color = "black", show.legend = FALSE) +
  #scale_size_manual(values = c(bait = 4, prey = 0.5)) +
  #scale_shape_manual(values = c(bait = 23, prey = 1)) +
  scale_shape_manual(values = c(PIV  = 0, RSV = 2, PIV_RSV = 19)) +
  scale_alpha_manual(values = c(PIV = 0.5, RSV = 0.5, PIV_RSV = 0.8)) + 
  coord_fixed() + theme_void()
p

BackupAsPDF(p)

#png (ScriptAndDatedFileName("TSNE_RSV_"))

```


```{r}
PIV <- fread ("~/UCSF/kroganlab/BenPolacco/roche/2023_06_05_PIV_NP_HeatDiffusion_data/2023_06_05_NP.allIntegrations.csv.gz")[omics == "ab_ph_ppi_rna_ub"]
RSV <- fread ("~/UCSF/kroganlab/BenPolacco/roche/2023_06_02_RSV_NetProp_HeatDiffusion_data/2023_06_05_NP.allIntegrations.csv.gz")[omics == "ab_ph_ppi_rna_ub"]
alan <- openxlsx::read.xlsx("/Users/ben/Downloads/VIROME_MASTER_TABLE_EXTERNAL.xlsx")
alan.shortlist <- openxlsx::read.xlsx("/Users/ben/Downloads/VIROME_MASTER_TABLE_EXTERNAL.xlsx", sheet = 2)
```

```{r}
PIV[symbol %in% alan.shortlist$X2] |> View()


```



