---
title: "18_12_24_MSAVisualizationScript"
author: "Martin Gordon"
date: "2024-12-18"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r}
library(data.table)
library(magrittr)
library(ggplot2)
library(ComplexHeatmap)
library(rjson)
library(stringr)
library(Biostrings)
#library(viridis)
```

color palettes...

```{r}
getQualitativePalette <- function(n) {

  col.pal <- c(
  "dodgerblue2", 
  "#E31A1C", # red
  "green4",
  "#6A3D9A", # purple
  "#FF7F00", # orange
  "black",
   "gold1",
  "skyblue2", 
  "#FB9A99", # lt pink
  "palegreen2",
  "#CAB2D6", # lt purple
  "#121111", # lt orange
  "gray70", 
  "khaki2",
  "maroon", 
  "orchid1", 
  "deeppink1",
   "blue1", 
  "steelblue4",
  "darkturquoise",
   "green1", 
   "yellow4", 
   "yellow3",
  "darkorange4",
   "brown"
  )
  return(col.pal[1:n])
}


#https://www.bioinformatics.nl/~berndb/aacolour.html Cinema
resColorsMulti = c(HKR = "#00FFFF",
              DE  = "#FF0000",
              STNQ = "#00FF00",
              AVLIM = "#BBBBBB",
              FWY = "#FF00FF",
              PG = "#996000",
              C ="#FFFF00",
              BZX = "grey25",
              "-" = "grey95")
```


Functions to lad the MSA
```{r}
loadAlnAsMatrix <- function (path){
  aln <- scan(path, character())
  aln.mat <- do.call(rbind, strsplit(aln, ""))
}



loadAnotDT <- function(afOutDir){
  fastaPath <- list.files(path = afOutDir, pattern = ".fasta", full.names = TRUE)
  print (fastaPath)
  stopifnot (`Unexpected number of fasta files found` = length(fastaPath) == 1 )
  fasta <- seqinr::read.fasta(fastaPath, seqtype ="AA" )
  anot.dt <- rbindlist( lapply(names(fasta), 
                               function(x)data.table(protein =  x,
                                                     aa = as.character(fasta[[x]]),
                                                     # this chain designation makes assumptions...
                                                     chain = LETTERS[which(names(fasta) == x)])))
  
  anot.dt[, protein := factor(protein, levels=unique(protein))] # enforce ordering
  anot.dt[, position := .SD[,.I], by = protein]
  anot.dt[, I := .I]
  
  # if available
  ccmpredFile <- list.files(path = afOutDir, pattern = ".ccmpred", full.names = TRUE)
  if (length(ccmpredFile) == 1){
    cc.mat <- loadCcmpredMat(ccmpredFile)
    a2bMat <- cc.mat[ anot.dt[, .I[chain == "A"]], anot.dt[, .I[chain == "B"]]]
    b2aMat <- cc.mat[ anot.dt[, .I[chain == "B"]], anot.dt[, .I[chain == "A"]]]
    
    anot.dt[chain == "A", bestCrossCCM := apply(a2bMat, 1, max, na.rm = TRUE)]
    anot.dt[chain == "B", bestCrossCCM := apply(b2aMat, 1, max, na.rm = TRUE)]
    
  }
  
  
  return (anot.dt[])
}

anot.dt <- loadAnotDT("~/Box/2024_03_PCMI_AF/af.pcmi/output/SKI__DCAF7/")
aln <- loadAlnAsMatrix("~/Box/2024_03_PCMI_AF/af.pcmi/output/SKI__DCAF7/features.msa.aln")
Heatmap(aln, cluster_rows = FALSE, cluster_columns = FALSE, col = resColors, name = "AAtype",
        column_labels = ifelse(anot.dt$position%%50 == 0, anot.dt$position, ""), column_names_side = "bottom")
```

```{r}
  # alphafold/common/residue_constants.py
  restypes = c(
    'A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P',
    'S', 'T', 'W', 'Y', 'V', 'x')
  
  msa[msa == 21] <- NA
  msa <- msa + 1
  
  
  .rmNA <- function(x)x[!is.na(x)]
  .singleSeq <- function(intCodes){
    paste0(restypes[.rmNA(intCodes)], collapse = "")
  }
  
  apply(msa, 1, .singleSeq)
}
msaMat2Seqs(a[1:2,])

Heatmap()
```


Load in the A3M matrix directly
```{r}
# supply the file path to read in the matrix
message('Reading in msa from path..')

readA3Mmatrix <-  function(a3mPath, sep=' '){
  
  message('Reading in a3m matrix from', a3mPath)
  a3m.mat <- fread(a3mPath, sep=sep) %>%  
  as.matrix()
  message('matrix dimensions: ' , paste(dim(a3m.mat), collapse=','))
  return(a3m.mat)
}

a3m.mat <- readA3Mmatrix(a3mPath = './data/O00631__P62328.msaOut.txt', sep=' ')
View(a3m.mat)
# AF3 feature encoding aplahfold/src/data/msa_features.py
# seems to be some redundancy in the code...
#   # AF3 _PROTEIN_TO_ID = {
#     'A': 0,
#     'B': 3,  # Same as D.
#     'C': 4,
#     'D': 3,
#     'E': 6,
#     'F': 13,
#     'G': 7,
#     'H': 8,
#     'I': 9,
#     'J': 20,  # Same as unknown (X).
#     'K': 11,
#     'L': 10,
#     'M': 12,
#     'N': 2,
#     'O': 20,  # Same as unknown (X).
#     'P': 14,
#     'Q': 5,
#     'R': 1,
#     'S': 15,
#     'T': 16,
#     'U': 4,  # Same as C.
#     'V': 19,
#     'W': 17,
#     'X': 20,
#     'Y': 18,
#     'Z': 6,  # Same as E.
#     '-': 21,
# }

  msa[msa == 21] <- NA
  msa <- msa + 1
  .rmNA <- function(x)x[!is.na(x)]
  .singleSeq <- function(intCodes){
    paste0(restypes[.rmNA(intCodes)], collapse = "")
  }
  apply(msa, 1, .singleSeq)
}


convertEncodingToResidueMatrix <-  function(a3mMat, removeNA=F){
  
  # AF2 alphafold/common/residue_constants.py follows same encoding as AF3 aplahfold/src/data/msa_features.py
  # seems to be some redundancy in the code...
  restypes = c('A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V', 'X', '-')
  
  #add  1 as python indexing
  a3mMat <- a3mMat + 1
  message('Converting numeric encoding to aa.residues...')
  
  .rmNA <- function(x)x[!is.na(x)]
  
  if(removeNA)
    message('Removing NA values')
  
   .singleSeq <- function(intCodes, removeNA){
     # recover the sequence per MSA record
     if (removeNA==T){
       paste0(restypes[.rmNA(intCodes)], collapse = "")
     } else {
       paste0(restypes[intCodes], collapse = "")
     }
   }

  #  get chr str of aa sequence
  resStr <- apply(a3mMat, 1, .singleSeq, removeNA=removeNA)
  # convert to chr vector
  resVec <- lapply(resStr, function(x) {unlist(strsplit(x, ''))} )
  resMat <- do.call(rbind, resVec) %>% 
     as.matrix()
  
  return(resMat)
}


# no RM seq
convertEncodingToResidueMatrix <-  function(a3mMat){
  
  # AF2 alphafold/common/residue_constants.py follows same encoding as AF3 aplahfold/src/data/msa_features.py
  # seems to be some redundancy in the code...
  restypes <-  c('A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V', 'X', '-')
  
  #add  1 as python indexing
  a3mMat <- a3mMat + 1
  message('Converting numeric encoding to aa.residues...')
  
  .rmNA <- function(x)x[!is.na(x)]
  .singleSeq <- function(intCodes){
       paste0(restypes[.rmNA(intCodes)], collapse = "")
   }

  #  get chr str of aa sequence
  resStr <- apply(a3mMat, 1, .singleSeq)
  # convert to chr vector
  resVec <- lapply(resStr, function(x) {unlist(strsplit(x, ''))} )
  resMat <- do.call(rbind, resVec) %>% 
     as.matrix()
  
  return(resMat)
}

# get the residue matrix.. lets remove the NA

res.mat <- convertEncodingToResidueMatrix(a3m.mat)
```

Preparing color encoding for the matrix
```{r}

cdt <- data.table(resids = names(resColorsMulti), color = resColorsMulti)
# this splits out the groups in the dt names to indiviudal records...
cdt <- cdt[, .(resid = unlist(strsplit(resids, ""))), by= color]

resColors = cdt$color %>%  setNames(cdt$resid)
```

Visualize the MSA to see the alignment; this looks good
```{r}
dim(res.mat)

rowF <- apply(res.mat, 1, function(x) sum(grepl('A', x)) != 256)
dim(res.mat[rowF,])
Heatmap(res.mat[1:160,1:75], 
        cluster_rows = FALSE, 
        cluster_columns = FALSE, 
        row_title = sprintf('%s MSA sequences', nrow(res.mat)),
        col = resColors, 
        name = "AAtype",
        column_names_side = "bottom")


Heatmap(res.mat[1:180,1:75], 
        cluster_rows = FALSE, 
        cluster_columns = FALSE, 
        row_title = sprintf('%s MSA sequences', nrow(res.mat)),
        col = resColors, 
        name = "AAtype",
        column_names_side = "bottom")


View(res.mat[1:40,1:75])
```
Need to get the annotation data; read this in from 
In JSON format, so read in these 


This is the pae plot.. should be able to tolerate multiple chains
*Todo* tidy the names of the output
```{r}
message('Read in the JSON files..')
confidences.json <- fromJSON(file='./data/p60484-1__o94832_confidences.json')
data.json <- fromJSON(file='./data/p60484-1__o94832_data.json')

message('getting protein chains')
# assuming the first is A, second is B etc...
prot.names <- strsplit(data.json$name, '__')[[1]]

message('Getting pae matrix')
# rbind pae vec to recover the matrix
pae.mat <- do.call(rbind, confidences.json$pae) %>% 
  as.matrix()


message('generating color palette for PAE')
chain.col.dt <- data.table(chainID = LETTERS[seq(1, length(prot.names))],
                           color = getQualitativePalette(n=length(prot.names)) )

chain.col.dt <- merge(chain.col.dt, data.table(chainID=unlist(confidences.json['token_chain_ids'])), by='chainID')

chainColVec <- chain.col.dt$color
names(chainColVec) <-  chain.col.dt$chainID


column_ha = HeatmapAnnotation(chainID=chain.col.dt$chainID, col=list(chainID=chainColVec))
row_ha = rowAnnotation(chainID = chain.col.dt$chainID, col=list(chainID=chainColVec))

hm <- Heatmap(pae.mat, 
        cluster_columns = F,
        cluster_rows = F,
        name = paste(data.json$name), 
        top_annotation = column_ha, 
        right_annotation = row_ha)

draw(hm)
```
Trim the tail of the MSA... get rid of this a padding... align the input sequences
Or just take the length of the input seq/ n residues etc, from the confidence and then use this to trim off the polyA tail of the matrix?
Maybe safer, as not relying on order of the chains..

```{r}
removeMSAPadding <- function(dataJson, resMat){
  
  message('Removing padding from MSA...\n')
  
  inputSeq <- lapply(seq_along(1:length(prot.names)), function(x){data.json$sequences[[x]]$protein$sequence}) %>% 
    paste0(collapse='')
  
  message('AF3 Query Sequence:\n', inputSeq)
  
  #get the top sequence from the MSA
  msaSeq <- paste0(resMat[1,],collapse='')

  match_idx <- matchPattern(inputSeq, msaSeq, max.mismatch = 0) %>% 
    as.data.table()
  
  # get the final indx, and get the residue
  return(resMat[,1:match_idx$end])
}

tidy.mat <- removeMSAPadding(dataJson = data.json, resMat = res.mat)
```
Now lets look at the tided protein

```{r}
message('Generating MSA plot..')

hm <- Heatmap(tidy.mat, 
        border=T,
        cluster_rows = FALSE, 
        cluster_columns = FALSE, 
        row_title = sprintf('%s MSA records', nrow(tidy.mat)),
        col = resColors, 
        top_annotation = column_ha, 
        column_split = chain.col.dt$chainID,
        name = "AAtype",
        column_names_side = "bottom")


```
Add the PLDDT score to the top of the MSA; lets see if MSA depth correlates with PLDDT score....

```{r}
data.json
```


Easier, just take the ncol of the res.mat 
```{r}


data.json %>%  View()





findResMatPadding <- function()

confidences.json %>% names()


strsplit(confidences.json['token_res_ids'], '')


unlist(confidences.json['token_res_ids'])

restypes <- c('A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V', 'X', '-')



paste0(restypes[.rmNA(intCodes)], collapse = "")




t(matrix(unlist(confidences.json['token_res_ids'])))

convertEncodingToResidueMatrix(t(matrix(unlist(confidences.json['token_res_ids']))))

convertEncodingToQuery <- function(a3mMat){
  
  # AF2 alphafold/common/residue_constants.py follows same encoding as AF3 aplahfold/src/data/msa_features.py
  # seems to be some redundancy in the code...
  restypes <-  c('A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V', 'X', '-')
  
  #add  1 as python indexing
  a3mMat <- a3mMat + 1
  message('Converting numeric encoding to aa.residues...')
  
  .rmNA <- function(x)x[!is.na(x)]
  .singleSeq <- function(intCodes){
       paste0(restypes[.rmNA(intCodes)], collapse = "")
   }

  #  get chr str of aa sequence
  resStr <- apply(a3mMat, 1, .singleSeq)
  # convert to chr vector
  resVec <- lapply(resStr, function(x) {unlist(strsplit(x, ''))} )
  resMat <- do.call(rbind, resVec) %>% 
     as.matrix()
  
  return(resMat)
}


convertEncodingToQuery(matrix(unlist(confidences.json['token_res_ids'])))[1:100]




length(unlist(confidences.json['token_res_ids']))


res.mat[1, 1:length(unlist(confidences.json['token_res_ids']))]
```



```{r}

seq_along(1:data.json['seqences'])

data.json %>% View()
data.json %>% names()

length(data.json[['sequences']][[1]][[1]])

lapply(seq_along(data.json['sequences']), function(x){
  
  print(x['id'])
  
})



# get the query sequences and if 




lapply(seq_along(1:length(prot.names)), function(x){
  

  prot.dt <- data.table(chainID=data.json$sequences[[x]]$protein$id,
             seq=data.json$sequences[[x]]$protein$sequence)
  
}) %>% rbindlist(idcol=FALSE)
       

inputSeq <- lapply(seq_along(1:length(prot.names)), function(x){
  
             seq=data.json$sequences[[x]]$protein$sequence
  
}) %>% paste0(collapse='')


inputSeq
paste0(res.mat[1,],collapse='')

data.json['sequences'][[1]]['protein']['id']
data.json['seqences'][1]
       
for (i in data.json['seqences']){
  
  if 
}

data.json$sequences[[2]]$protein$id
data.json$sequences[[2]]$protein$sequence


View(a3m.mat)

31+44 #75 aa in total?
```


```{r} 
# function to remove MSA padding; read in the orgin
#removeMSAPadding <- 

# this is going to be our row and column annotation
chainIDs <- confidences.json$token_chain_ids

# rbind pae vec to recover the matrix
pae.mat <- do.call(rbind, confidences.json$pae) %>% 
  as.matrix()


data.json$name
```


