---
title: "071223.HybridGenome.Creation"
author: "Martin Gordon"
date: "2023-12-07"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Create a hybrid genome for Jyoti RNAseq project
Want to map to the viral strains using the host genome as decoy

Will use GRCh-38 ensembl release 101 primary sequence for project legacy

This release has multi-placed sequences in PAR regions of ChrX ChrY, so not suitable to consider reads in this region
See below blog on choosing appropriate reference:
https://lh3.github.io/2017/11/13/which-human-reference-genome-to-use

```{r}
library(data.table)
library(rtracklayer)
library(magrittr)

```
```{bash eval=FALSE}
#mkdir -p ./{docs,data,script,output}
```

# download the host genome in fasta format and annotation files 
just downloading from ensembl ftp site as server closer atm....

```{bash eval=FALSE}
#wget 'ftp://ftp.ensembl.org/pub/release-101/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa.gz' ./data #soft masked for general usability. may need to create bed file of masked regions for downstream filtering...
#wget 'ftp://ftp.ensembl.org/pub/release-101/gtf/homo_sapiens/Homo_sapiens.GRCh38.101.gtf.gz' ./data
#wget 'ftp://ftp.ensembl.org/pub/release-101/gff3/homo_sapiens/Homo_sapiens.GRCh38.101.gff3.gz' ./data

```

# download the viral genomes (InF Brisbane and InFB Victoria) and their gffs
didnt order 
### Edit GFFs for joining   
```{r}
#CoV2
brisbane_gff <-  readGFF("./data/brisbane.gff", 
                   version = 3)
victoria_gff <-  readGFF("./data/victoria.gff3", 
                   version = 3)

brisbane_gff <-  subset(brisbane_gff,  type == 'gene')
victoria_gff <-  subset(victoria_gff,  type == 'gene')

export.gff3(brisbane_gff, "./data/brisbane.cleaned.gff")
export.gff3(victoria_gff, "./data/victoria.cleaned.gff")
#all the 9 cols are the same so should be fine
```

### Generate additional files   
gffread version 
```{bash eval=FALSE}
conda activate gffread

gffread  \
'./data/victoria.cleaned.gff' \
-T \
-F \
 --gene2exon \
-o './data/victoria.cleaned.gtf' #make gtf from gff

gffread  \
--gene2exon \
'./data/victoria.cleaned.gff' \
--bed \
-o './data/victoria.cleaned.bed' #make BED from gff

# CoV2
gffread  \
'./data/brisbane.cleaned.gff' \
-T \
-F \
 --gene2exon \
-o './data/brisbane.cleaned.gtf' #make gtf from gff

gffread  \
--gene2exon \
'./data/brisbane.cleaned.gff' \
--bed \
-o './data/brisbane.cleaned.bed' #make BED from gff


#join BED files
cat '/projects/p30623/reference_genomes/GRCh38_hardcopy/Homo_sapiens.GRCh38.100.bed' \
'/projects/p30623/reference_genomes/SARS_CoV_2_NC_045512.2/GCF_009858895.2_ASM985889v3_genomic_cleaned.bed' \
'/projects/p30623/reference_genomes/CF_001343785.1/GCF_001343785.1_ViralMultiSegProj274766_genomic_cleaned.bed' > 'human_viral_pneumonia.bed'
#join GFFs 
cat '/projects/p30623/reference_genomes/GRCh38_hardcopy/Homo_sapiens.GRCh38.100.gff3' \
'/projects/p30623/reference_genomes/SARS_CoV_2_NC_045512.2/GCF_009858895.2_ASM985889v3_genomic_cleaned.gff' \
'/projects/p30623/reference_genomes/CF_001343785.1/GCF_001343785.1_ViralMultiSegProj274766_genomic_cleaned.gff' > 'human_viral_pneumonia.gff'
```
